[![GitHub Downloads](https://img.shields.io/github/downloads/EI-CoreBioinformatics/mikado/total.svg?style=social&logo=github&label=download)](https://github.com/EI-CoreBioinformatics/mikado/releases)
[![Release](https://img.shields.io/github/release/EI-CoreBioinformatics/mikado.svg)](https://github.com/EI-CoreBioinformatics/mikado/releases)
[![PyPI](https://img.shields.io/pypi/v/mikado.svg?style=flat)](https://pypi.python.org/pypi/mikado)
[![Build Status](https://github.com/EI-CoreBioinformatics/mikado/workflows/Mikado/badge.svg)](https://github.com/EI-CoreBioinformatics/mikado/actions/workflows/python-package.yml)

# Mikado - pick your transcript: a pipeline to determine and select the best RNA-Seq prediction

Mikado is a lightweight Python3 pipeline to identify the most useful or “best” set of transcripts from multiple transcript assemblies. Our approach leverages transcript assemblies generated by multiple methods to define expressed loci, assign a representative transcript and return a set of gene models that selects against transcripts that are chimeric, fragmented or with short or disrupted CDS. Loci are first defined based on overlap criteria and each transcript therein is scored based on up to 50 available metrics relating to ORF and cDNA size, relative position of the ORF within the transcript,  UTR length and presence of multiple ORFs. Mikado can also utilize blast data to score transcripts based on proteins similarity and to identify and split chimeric transcripts. Optionally, junction confidence data as provided by [Portcullis][Portcullis] can be used to improve the assessment. The best-scoring transcripts are selected as the primary transcripts of their respective gene loci; additionally, Mikado can bring back other valid splice variants that are compatible with the primary isoform.

Mikado uses GTF or GFF files as mandatory input. Non-mandatory but highly recommended input data can be generated by obtaining a set of reliable splicing junctions with Portcullis_, by locating coding ORFs on the transcripts using either [Transdecoder][Transdecoder] or [Prodigal][Prodigal], and by obtaining homology information through either [BLASTX][Blast+] or [DIAMOND][DIAMOND].

Our approach is amenable to include sequences generated by *de novo* Illumina assemblers or reads generated from long read technologies such as Pacbio.

Extended documentation is hosted on ReadTheDocs: http://mikado.readthedocs.org/

## Contents
- [Installation](#installation)
  - [Docker Installation](#docker-installation)
  - [Singularity Installation](#singularity-installation)
  - [Conda Installation](#condamambamanual-installation)
    
## Installation

### Docker Installation
Mikado can be installed with docker. If you don't have docker, please install [docker](https://docs.docker.com/get-docker/) first. Then you can pull the docker image with mikado installed
```console
VERSION=2.3.5rc2
docker run gemygk/mikado:v${VERSION} mikado -h
```

### Singularity Installation
Mikado can be installed with singularity. If you don't have singularity, please install [singularity](https://docs.sylabs.io/guides/3.9/user-guide/quick_start.html#quick-installation-steps) first. Then you can pull the singularity image with mikado installed.
```console
VERSION=2.3.5rc2
singularity exec docker://gemygk/mikado:v${VERSION} mikado -h
```
Or, we can build and run a singularity image
```console
# 1. Create a Singularity definition file

$ cat mikado-2.3.5rc2.def
bootstrap: docker
from: gemygk/mikado:v2.3.5rc2

# Build the Singularity image
$ sudo singularity build mikado-2.3.5rc2.sif mikado-2.3.5rc2.def

# Execute Mikado
$ singularity exec mikado-2.3.5rc2_CBG.sif mikado -h
usage: Mikado [-h] [--version] {configure,prepare,serialise,pick,compare,util} ...

Mikado is a program to analyse RNA-Seq data and determine the best transcript for each locus in accordance to user-specified criteria.

optional arguments:
  -h, --help            show this help message and exit
  --version             Print Mikado current version and exit.

Components:
  {configure,prepare,serialise,pick,compare,util}
                        These are the various components of Mikado:
    configure           This utility guides the user through the process of creating a configuration file for Mikado.
    prepare             Mikado prepare analyses an input GTF file and prepares it for the picking analysis by sorting its transcripts and performing some simple consistency checks.
    serialise           Mikado serialise creates the database used by the pick program. It handles Junction and ORF BED12 files as well as BLAST XML results.
    pick                Mikado pick analyses a sorted GTF/GFF files in order to identify its loci and choose the best transcripts according to user-specified criteria. It is dependent on files produced by the "prepare" and "serialise"
                        components.
    compare             Mikado compare produces a detailed comparison of reference and prediction files. It has been directly inspired by Cufflinks's cuffcompare and ParsEval.
    util                Miscellaneous utilities
```

### Conda/Mamba/Manual Installation
Mikado can be installed with conda. If you don't have conda, please install [mamba](https://github.com/conda-forge/miniforge?tab=readme-ov-file#miniforge-pypy3) first. Then you can create a new environment with mikado installed.

Install mamba with PyPy 3.9 in the base environment (https://github.com/conda-forge/miniforge?tab=readme-ov-file#miniforge-pypy3)

Replace `/path/to` with your installation directory when following the steps below:

```console
/path/to/src
[src]$ wget -c https://github.com/conda-forge/miniforge/releases/latest/download/Miniforge-pypy3-Linux-x86_64.sh
[src]$ bash Miniforge-pypy3-Linux-x86_64.sh
```

I have installed the base to `/path/to/x86_64/` location

If you have chosen to not have conda modify your shell scripts at all, to activate conda's base environment in your current shell session, please do:

```console
/path/to/src
[src]$ eval "$(/path/to/x86_64/bin/conda shell.bash hook)"
```

Install Git

```console
/path/to/src
(base) [src]$ mamba install -y git
```

Clone mikado
```console
/path/to/src
(base) [src]$ git clone git@github.com:EI-CoreBioinformatics/mikado.git
(base) [src]$ cd mikado
```

Install Mikado dependencies
```console
/path/to/src/mikado
(base) [mikado]$ mamba env create -f environment.yml --prefix /path/to/x86_64/envs/mikado_env
```

Activate mikado_env
```console
/path/to/src/mikado
(base) [mikado]$ conda activate mikado_env
(mikado_env) [mikado]$
```

Do checks if all dependencies are installed. A full list of library dependencies can be found in the file ``requirements.txt``
```console
/path/to/src/mikado
(mikado_env) [mikado]$ pip3 install wheel
(mikado_env) [mikado]$ pip3 install -r requirements.txt
(mikado_env) [mikado]$ pip3 install Cython
```
- For the above commands - wheel, requirements.txt, Cython should all have the status 'Requirement already satisfied'


We need gcc for bdist_wheel (tested on gcc v5.2.0, v9.4.0)

```console
/path/to/src/mikado
(mikado_env) [mikado]$ python3 setup.py bdist_wheel
(mikado_env) [mikado]$ pip3 install dist/*.whl
```
Now that installation is complete, run Mikado help
```console
/path/to/src/mikado
(mikado_env) [mikado]$ mikado -h
usage: Mikado [-h] [--version] {configure,prepare,serialise,pick,compare,util} ...

Mikado is a program to analyse RNA-Seq data and determine the best transcript for each locus in accordance to user-specified criteria.

optional arguments:
  -h, --help            show this help message and exit
  --version             Print Mikado current version and exit.

Components:
  {configure,prepare,serialise,pick,compare,util}
                        These are the various components of Mikado:
    configure           This utility guides the user through the process of creating a configuration file for Mikado.
    prepare             Mikado prepare analyses an input GTF file and prepares it for the picking analysis by sorting its transcripts and performing some simple consistency checks.
    serialise           Mikado serialise creates the database used by the pick program. It handles Junction and ORF BED12 files as well as BLAST XML results.
    pick                Mikado pick analyses a sorted GTF/GFF files in order to identify its loci and choose the best transcripts according to user-specified criteria. It is dependent on files produced by the "prepare" and "serialise"
                        components.
    compare             Mikado compare produces a detailed comparison of reference and prediction files. It has been directly inspired by Cufflinks's cuffcompare and ParsEval.
    util                Miscellaneous utilities
```



You can verify the correctness of the installation with the unit tests (*outside of the source folder*, as otherwise Python will get confused and try to use the `Mikado` source folder instead of the system installation):

    python -c "import Mikado; Mikado.test(); Mikado.test(label='slow')"


### Additional dependencies

Mikado by itself does require only the presence of a database solution, such as SQLite (although we do support MySQL and PostGRESQL as well).
However, the Daijin pipeline requires additional programs to run.

For driving Mikado through Daijin, the following programs are required:

- [DIAMOND][DIAMOND] or [Blast+][Blast+] to provide protein homology. DIAMOND is preferred for its speed.
- [Prodigal][Prodigal] or [Transdecoder][Transdecoder] to calculate ORFs. The versions of Transdecoder that we tested scale poorly in terms of runtime and disk usage, depending on the size of the input dataset. Prodigal is much faster and lighter, however, the data on our paper has been generated through Transdecoder - not Prodigal. Currently we set Prodigal as default.
- Mikado also makes use of a dataset of RNA-Seq high-quality junctions. We are using [Portcullis][Portcullis] to calculate this data alongside the alignments and assemblies.

If you plan to generate the alignment and assembly part as well through Daijin, the pipeline requires the following:

- SAMTools
- If you have short-read RNA-Seq data:
  - At least one short-read RNA-Seq aligner, choice between [GSNAP], [GMAP][GMAP], [STAR][STAR], [TopHat2][TopHat2], [HISAT2][HISAT2]
  - At least one RNA-Seq assembler, choice between [StringTie][StringTie], [Trinity][Trinity], [Cufflinks], [CLASS2][CLASS2]. Trinity additionally requires [GMAP][GMAP].
  - [Portcullis][Portcullis] is optional, but highly recommended to retrieve high-quality junctions from the data
- If you have long-read RNA-Seq data:
  - At least one long-read RNA-Seq aligner, current choice between [STAR][STAR] and [GMAP][GMAP]

## Development guide

We provide source trail files ([https://www.sourcetrail.com/](https://www.sourcetrail.com/)) to aid in development.
As required by the SourceTrail application, these files are present in the master directory, as "Mikado.srctrl*".

## Citing Mikado

If you use Mikado in your work, please consider to cite:

> Venturini L., Caim S., Kaithakottil G., Mapleson D.L., Swarbreck D. Leveraging multiple transcriptome assembly methods for improved gene structure annotation. GigaScience, Volume 7, Issue 8, 1 August 2018, giy093, [doi:10.1093/gigascience/giy093](https://doi.org/10.1093/gigascience/giy093)

If you also use Portcullis to provide reliable junctions to Mikado, either independently or as part of the Daijin pipeline, please consider to cite:

> Mapleson D.L., Venturini L., Kaithakottil G., Swarbreck D. Efficient and accurate detection of splice junctions from RNAseq with Portcullis. GigaScience, Volume 7, Issue 12, 12 December 2018, giy131, [doi:10.1093/gigascience/giy131](https://doi.org/10.1093/gigascience/giy131)

[Portcullis]: https://github.com/maplesond/portcullis
[Blast+]: https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download
[Transdecoder]: https://github.com/TransDecoder/TransDecoder/
[Prodigal]: https://github.com/hyattpd/Prodigal
[DIAMOND]: https://github.com/bbuchfink/diamond/
[GMAP]: http://research-pub.gene.com/gmap/
[STAR]: https://github.com/alexdobin/STAR
[TopHat2]: https://ccb.jhu.edu/software/tophat/index.shtml
[HISAT2]: http://ccb.jhu.edu/software/hisat2
[StringTie]: https://ccb.jhu.edu/software/stringtie/
[Trinity]: https://github.com/trinityrnaseq/trinityrnaseq
[CLASS2]: http://ccb.jhu.edu/people/florea/research/CLASS2/
