<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>6.2. The Daijin pipeline for driving Mikado &mdash; Mikado 1.0.0b4 documentation</title>
    
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0.0b4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Mikado 1.0.0b4 documentation" href="../index.html" />
    <link rel="up" title="4. Usage" href="../Usage.html" />
    <link rel="next" title="6.3. Mikado miscellaneous scripts" href="Utilities.html" />
    <link rel="prev" title="6.1. Compare" href="Compare.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Utilities.html" title="6.3. Mikado miscellaneous scripts"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Compare.html" title="6.1. Compare"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Mikado 1.0.0b4 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../Usage.html" accesskey="U">4. Usage</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="assemble-pipeline"></span><div class="section" id="the-daijin-pipeline-for-driving-mikado">
<span id="daijin"></span><h1>6.2. The Daijin pipeline for driving Mikado<a class="headerlink" href="#the-daijin-pipeline-for-driving-mikado" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="http://snakemake.bitbucket.org"><img alt="snake_badge" src="https://img.shields.io/badge/snakemake-≥3.5.2-brightgreen.svg?style=flat-square)]" /></a></p>
<p>No emperor or empress can lead its nation without a trusty chancellor to help him or her in organising the bureaucracy. Daijin, the Japanese minister, has the same role in Mikado - it smooths the path to go from a collection of read inputs (both RNA-Seq or long reads) to a polished transcriptome assembly. The pipeline is based on <a class="reference external" href="https://bitbucket.org/snakemake/snakemake/wiki/Home">Snakemake</a> <a class="reference internal" href="../References.html#snake" id="id1">[Snake]</a>; while Snakemake can support any scheduling system, our pipeline manager currently supports only three (SLURM, PBS and LSF), plus any DRMAA-compliant batch submission system. Other schedulers can be added upon request.</p>
<div class="admonition hint">
<p class="first admonition-title">Hint</p>
<p class="last">It is possible to launch the two steps of the pipeline directly with Snakemake, using the snakefiles located in Mikado.daijin: <a class="reference download internal" href="../_downloads/tr.snakefile" download=""><code class="xref download docutils literal"><span class="pre">tr.snakefile</span></code></a> for the first step, and <a class="reference download internal" href="../_downloads/mikado.snakefile" download=""><code class="xref download docutils literal"><span class="pre">mikado.snakefile</span></code></a> for the second.</p>
</div>
<div class="section" id="configure">
<span id="daijin-configure"></span><h2>6.2.1. Configure<a class="headerlink" href="#configure" title="Permalink to this headline">¶</a></h2>
<p>This utility creates the configuration file that will drive Daijin, in <a class="reference external" href="http://www.yaml.org/spec/1.2/spec.html">YAML</a> format. The file will need to be edited to</p>
</div>
<div class="section" id="assemble">
<span id="daijin-assemble"></span><h2>6.2.2. Assemble<a class="headerlink" href="#assemble" title="Permalink to this headline">¶</a></h2>
<p>In the first step of the pipeline, Daijin will perform the following operations for each of the read datasets provided:</p>
<ol class="arabic simple">
<li>Create the necessary indices for each of the aligner programs requested.</li>
<li>Align the read dataset using all the different tools requested, in all the possible combinations of parameters requested.
* For example, it is possible to ask each dataset to be aligned twice with TopHat2 - once with the &#8220;micro-exon&#8221; mode activated, the second time without. Both alignments will be run independently.
* It is possible to specify which datasets are strand-specific and which are not, and moreover, it is possible to specify the kind of strand-specificity (fr-secondstrand, fr-firststrand).</li>
<li>Call all the reliable junctions across the alignments using <a class="reference external" href="https://github.com/maplesond/portcullis">Portcullis</a>.</li>
<li>Create the statistics for the assembly using <code class="docutils literal"><span class="pre">samtools</span> <span class="pre">stat</span></code>, and merge them together in a single file.</li>
<li>Assemble each alignment with all the tools requested, in all the parameter combinations desired.</li>
<li>Call the statistics on each assembly using <a class="reference internal" href="Utilities.html#stat-command"><span class="std std-ref">mikado util stats</span></a>, and merge them together in a single file.</li>
<li>Create the configuration file for Mikado.</li>
</ol>
<p>So during this first step Daijin will go from raw reads files to multiple assemblies, and configure Mikado for the second step.</p>
<div class="topic">
<p class="topic-title first">Assembly pipeline, as driven by Daijin</p>
<div class="figure align-center" id="id9" style="width: 100%">
<a class="reference internal image-reference" href="../_images/daijin_assemble.svg"><img alt="../_images/daijin_assemble.svg" src="../_images/daijin_assemble.svg" /></a>
<p class="caption"><span class="caption-text">Example of a pipeline to assemble a single paired-end read dataset using three aligners (STAR <a class="reference internal" href="../References.html#star" id="id2">[STAR]</a>, Hisat <a class="reference internal" href="../References.html#hisat" id="id3">[Hisat]</a>, TopHat2 <a class="reference internal" href="../References.html#tophat2" id="id4">[TopHat2]</a> ) and two different RNA-Seq assemblers (StringTie <a class="reference internal" href="../References.html#stringtie" id="id5">[StringTie]</a>, CLASS2 <a class="reference internal" href="../References.html#class2" id="id6">[Class2]</a> ). Reliable junctions from the three alignments are called and merged together using <a class="reference external" href="https://github.com/maplesond/portcullis">Portcullis</a>.</span></p>
</div>
</div>
<p>Usage:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ daijin assemble --help
usage: daijin assemble [-h] [-c HPC_CONF] [--jobs N] [--cores [N]]
                       [--threads N] [--no_drmaa] [--rerun-incomplete]
                       [--forcerun TARGET [TARGET ...]] [--detailed-summary]
                       [--list] [--dag]
                       config

positional arguments:
  config                Configuration file to use for running the transcript
                        assembly pipeline.

optional arguments:
  -h, --help            show this help message and exit
  -c HPC_CONF, --hpc_conf HPC_CONF
                        Configuration file that allows the user to override
                        resource requests for each rule when running under a
                        scheduler in a HPC environment.
  --jobs N, -J N        Maximum number of cluster jobs to execute
                        concurrently.
  --cores [N], -C [N]   Use at most N cores in parallel (default: 1000).
  --threads N, -t N     Maximum number of threads per job. Default: None (set
                        in the configuration file)
  --no_drmaa, -nd       Use this flag if you wish to run without DRMAA, for
                        example, if running on a HPC and DRMAA is not
                        available, or if running locally on your own machine
                        or server.
  --rerun-incomplete, --ri
                        Re-run all jobs the output of which is recognized as
                        incomplete.
  --forcerun TARGET [TARGET ...], -R TARGET [TARGET ...]
                        Force the re-execution or creation of the given rules
                        or files. Use this option if you changed a rule and
                        want to have all its output in your workflow updated.
  --detailed-summary, -D
                        Print detailed summary of all input and output files
  --list, -l            List resources used in the workflow
  --dag                 Do not execute anything and print the redirected
                        acylic graph of jobs in the dot language.
</pre></div>
</div>
</div>
<div class="section" id="mikado">
<span id="daijin-mikado"></span><h2>6.2.3. Mikado<a class="headerlink" href="#mikado" title="Permalink to this headline">¶</a></h2>
<p>In this step, the Daijin manager will execute all the steps necessary to perform Mikado on the desired inputs. The manager will execute the following steps:</p>
<ol class="arabic simple">
<li>Merge all the input assemblies together using <a class="reference internal" href="Prepare.html#prepare"><span class="std std-ref">Mikado prepare</span></a></li>
<li>Execute <a class="reference external" href="https://github.com/TransDecoder/TransDecoder">TransDecoder</a> <a class="reference internal" href="../References.html#trinity" id="id7">[Trinity]</a> on the transcript sequences, to retrieve their ORFs.</li>
<li>Split the FASTA file in as many chunks as specified during configuration, and analyse them separately</li>
<li>Execute <a class="reference external" href="https://www.ncbi.nlm.nih.gov/books/NBK279690/">BLASTX+</a> <a class="reference internal" href="../References.html#blastplus" id="id8">[Blastplus]</a> on the splitted FASTAs, creating BLAST XML outputs.</li>
<li>Run <a class="reference internal" href="Serialise.html#serialise"><span class="std std-ref">Mikado serialise</span></a> to load the BLAST results, TransDecoder ORFs, and portcullis junctions into a single database.</li>
<li>Run <a class="reference internal" href="Pick.html#pick"><span class="std std-ref">Mikado pick</span></a> on the data, in the selected modes.</li>
<li>Collate and collapse the statistics for each of the filtered assemblies.</li>
</ol>
<div class="topic">
<p class="topic-title first">Mikado pipeline, as driven by Daijin</p>
<div class="figure align-center" id="id10">
<a class="reference internal image-reference" href="../_images/daijin_mikado.svg"><img alt="../_images/daijin_mikado.svg" src="../_images/daijin_mikado.svg" /></a>
<p class="caption"><span class="caption-text">Example of a typical Mikado pipeline. In this case the number of chunks for BLAST is limited - 10 - but we advise to increase this number for big datasets.</span></p>
</div>
</div>
<p>Command line usage:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ daijin mikado --help
usage: daijin mikado [-h] [-c HPC_CONF] [--jobs N] [--cores [N]] [--threads N]
                     [--no_drmaa] [--rerun-incomplete]
                     [--forcerun TARGET [TARGET ...]] [--detailed-summary]
                     [--list] [--dag]
                     config

positional arguments:
  config                Configuration file to use for running the Mikado step
                        of the pipeline.

optional arguments:
  -h, --help            show this help message and exit
  -c HPC_CONF, --hpc_conf HPC_CONF
                        Configuration file that allows the user to override
                        resource requests for each rule when running under a
                        scheduler in a HPC environment.
  --jobs N, -J N        Maximum number of cluster jobs to execute
                        concurrently.
  --cores [N], -C [N]   Use at most N cores in parallel (default: 1000).
  --threads N, -t N     Maximum number of threads per job. Default: None (set
                        in the configuration file)
  --no_drmaa, -nd       Use this flag if you wish to run without DRMAA, for
                        example, if running on a HPC and DRMAA is not
                        available, or if running locally on your own machine
                        or server.
  --rerun-incomplete, --ri
                        Re-run all jobs the output of which is recognized as
                        incomplete.
  --forcerun TARGET [TARGET ...], -R TARGET [TARGET ...]
                        Force the re-execution or creation of the given rules
                        or files. Use this option if you changed a rule and
                        want to have all its output in your workflow updated.
  --detailed-summary, -D
                        Print detailed summary of all input and output files
  --list, -l            List resources used in the workflow
  --dag                 Do not execute anything and print the redirected
                        acylic graph of jobs in the dot language.
</pre></div>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">If you have already created some assemblies and wish to analyse them with Daijin, it is also possible to <a class="reference internal" href="Configure.html#configure"><span class="std std-ref">configure Mikado externally</span></a> and use the resulting configuration file to guide Daijin.</p>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/mikado-logo.jpg" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">6.2. The Daijin pipeline for driving Mikado</a><ul>
<li><a class="reference internal" href="#configure">6.2.1. Configure</a></li>
<li><a class="reference internal" href="#assemble">6.2.2. Assemble</a></li>
<li><a class="reference internal" href="#mikado">6.2.3. Mikado</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Compare.html"
                        title="previous chapter">6.1. Compare</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Utilities.html"
                        title="next chapter">6.3. Mikado miscellaneous scripts</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/Usage_dir/Daijin.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Utilities.html" title="6.3. Mikado miscellaneous scripts"
             >next</a> |</li>
        <li class="right" >
          <a href="Compare.html" title="6.1. Compare"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Mikado 1.0.0b4 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../Usage.html" >4. Usage</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Venturini Luca, Caim Shabhonam, Mapleson Daniel, Kaithakottil Gemy George, Swarbreck David.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.5.
    </div>
  </body>
</html>