<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>5.4. Mikado pick &mdash; Mikado 1.0.0b4 documentation</title>
    
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0.0b4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Mikado 1.0.0b4 documentation" href="../index.html" />
    <link rel="up" title="4. Usage" href="../Usage.html" />
    <link rel="next" title="6.1. Compare" href="Compare.html" />
    <link rel="prev" title="5.3. Mikado serialise" href="Serialise.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Compare.html" title="6.1. Compare"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Serialise.html" title="5.3. Mikado serialise"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Mikado 1.0.0b4 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../Usage.html" accesskey="U">4. Usage</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="mikado-pick">
<span id="pick"></span><h1>5.4. Mikado pick<a class="headerlink" href="#mikado-pick" title="Permalink to this headline">¶</a></h1>
<p>This is the final stage of the pipeline, and the one which actually applies the <a class="reference internal" href="../Introduction.html#introduction"><span class="std std-ref">core algorithms</span></a> to identify the loci and select the best transcripts.
As such, it is the</p>
<div class="section" id="usage">
<h2>5.4.1. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">mikado</span> <span class="pre">pick</span></code> allows to modify some of the parameters regarding the run at runtime. However, some sections - such as most of the settings regarding alternative splicing - are left untouched by the utility, and are best modified by editing the <a class="reference internal" href="Configure.html#configure"><span class="std std-ref">configuration file itself</span></a>. The available parameters are as follows:</p>
<ul>
<li><p class="first"><em>json-conf</em>: required. This is the configuration file created in the <a class="reference internal" href="Configure.html#configure"><span class="std std-ref">first step</span></a> of the pipeline.</p>
</li>
<li><p class="first"><em>gff</em>; optionally, it is possible to point Mikado prepare to the GTF it should use here on the command line. This file should be the output of the <a class="reference internal" href="Prepare.html#prepare"><span class="std std-ref">preparation step</span></a>. Please note that this file should be in GTF format, sorted by chromosome and position; if that is not the case, Mikado will fail.</p>
</li>
<li><p class="first"><em>db</em>: Optionally, it is possible to specify the database to Mikado on the command line, rather than on the configuration file. Currently, this option <em>supports SQLite databases only</em>.</p>
</li>
<li><dl class="first docutils">
<dt>Options related to how Mikado will treat the data:</dt>
<dd><ul class="first last simple">
<li><em>intron_range</em>: this option expects a couple of positive integers, in ascending order, indicating the 98% CI where most intron lengths should fall into. Gene models with introns whose lengths fall outside of this range might be penalized, depending on the scoring system used. If uncertain, it is possible to use the <a class="reference internal" href="Utilities.html#stat-command"><span class="std std-ref">included stats utility</span></a> on the gene annotation of a closely related species.</li>
<li><em>purge</em>: flag. If set, Mikado will not just identify putative fragments - it will completely exclude them from the output.</li>
<li><em>flank</em>: for the purposes of identifying fragments, it is useful to consider together loci which are not necessarily overlapping but that lying relatively near on the genome sequence. This parameter (a positive integer) specifies the maximum distance for Mikado for gathering data together for this purpose.</li>
<li><em>mode</em>: how Mikado will treat BLAST and ORF data in the presence of putative chimeras. See the <a class="reference internal" href="Configure.html#chimera-splitting"><span class="std std-ref">relevant section in the configuration page</span></a> for details.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Options regarding the output files:</dt>
<dd><ul class="first last simple">
<li><em>output-dir</em>: Output directory. By default, Mikado will write all files and the log on the current directory.</li>
<li><em>loci_out</em>: required. This it the main output file, in GFF format.</li>
<li><em>prefix</em>: this ID will be prefixed to all gene and transcript models. IN general, IDs will be of the form &#8220;&lt;prefix&gt;.&lt;chromosome&gt;&lt;progressive ID&gt;&#8221;. Default: Mikado.</li>
<li><em>source</em>: source field prefix for the output files. Useful for eg loading Mikado runs into <a class="reference external" href="http://genomearchitect.org/">WebApollo</a> <a class="reference internal" href="../References.html#apollo" id="id1">[Apollo]</a>.</li>
<li><em>no_cds</em>: if present, this flg will indicate to Mikado not to print out the CDS of selected models but only their transcript structures.</li>
<li><em>subloci_out</em>: If requested, Mikado can output the data regarding the first intermediate step, ie the <em>subloci</em>. See the <a class="reference internal" href="../Introduction.html#introduction"><span class="std std-ref">introduction</span></a> for details.</li>
<li><em>monoloci_out</em>: If requested, Mikado can output the data regarding the second intermediate step, ie the <em>monosubloci</em>. See the <a class="reference internal" href="../Introduction.html#introduction"><span class="std std-ref">introduction</span></a> for details.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Options regarding the resources to be used:</dt>
<dd><ul class="first last simple">
<li><em>procs</em>: number of processors to use.</li>
<li><em>start-method</em>: multiprocessing start method. See the <a class="reference internal" href="Configure.html#scheduler-multiprocessing"><span class="std std-ref">explanation on Python multiprocessing</span></a></li>
<li><em>single</em>: flag. If present, multiprocessing will be disabled.</li>
<li><em>shared-memory</em>: flag, available on Unix systems only. If set, Mikado will try to copy the SQLite database in RAM. It might provide a small boost if disk access is a limitation. Ineffective with databases other than SQLite.</li>
<li><em>shared-memory-db</em>: if the database has already been copied in memory, its new location can be given with this argument. Useful to have multiple Mikado picks share the same DB.</li>
<li><em>preload</em>: flag. If present, Mikado will load the database in memory before execution. <strong>Discouraged unless the database size is quite small.</strong></li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Options regarding logging:</dt>
<dd><ul class="first last simple">
<li><em>log</em>: name of the log file. By default, &#8220;pick.log&#8221;</li>
<li><em>verbose</em>: sets the log level to DEBUG. Please be advised that the debug mode is <strong>extremely</strong> verbose and is bestly invoked only for real, targeted debugging sessions.</li>
<li><em>noverbose</em>: sets the log level to ERROR. If set, in most cases, the log file will be practically empty.</li>
<li><em>log-level</em>: this flag directly sets the log level. Available values: DEBUG, INFO, WARNING, ERROR.</li>
</ul>
</dd>
</dl>
</li>
</ul>
<p>Usage:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ mikado pick --help
usage: Mikado pick [-h] [--start-method {fork,spawn,forkserver}] [-p PROCS]
                   --json-conf JSON_CONF [-i INTRON_RANGE INTRON_RANGE]
                   [--subloci_out SUBLOCI_OUT] [--monoloci_out MONOLOCI_OUT]
                   [--loci_out LOCI_OUT] [--prefix PREFIX] [--no_cds]
                   [--source SOURCE] [--flank FLANK] [--purge] [-shm]
                   [-shmdb SHM_DB] [--preload] [-db SQLITE_DB]
                   [-od OUTPUT_DIR] [--single] [-l LOG] [-v | -nv]
                   [-lv {DEBUG,INFO,WARN,ERROR,CRITICAL}]
                   [--mode {nosplit,stringent,lenient,permissive,split}]
                   [gff]

Mikado pick analyses a sorted GTF/GFF files in order to identify its loci and
choose the best transcripts according to user-specified criteria. It is
dependent on files produced by the &quot;prepare&quot; and &quot;serialise&quot; components.

positional arguments:
  gff

optional arguments:
  -h, --help            show this help message and exit
  --start-method {fork,spawn,forkserver}
                        Multiprocessing start method. (default: None)
  -p PROCS, --procs PROCS
                        Number of processors to use. Default: look in the
                        configuration file (1 if undefined) (default: None)
  --json-conf JSON_CONF
                        JSON/YAML configuration file for scoring transcripts.
                        (default: None)
  -i INTRON_RANGE INTRON_RANGE, --intron-range INTRON_RANGE INTRON_RANGE
                        Range into which intron lengths should fall, as a
                        couple of integers. Transcripts with intron lengths
                        outside of this range will be penalised. Default: (60,
                        900) (default: None)
  --subloci_out SUBLOCI_OUT
  --monoloci_out MONOLOCI_OUT
  --loci_out LOCI_OUT   This output file is mandatory. If it is not specified
                        in the configuration file, it must be provided here.
                        (default: None)
  --prefix PREFIX       Prefix for the genes. Default: Mikado (default: None)
  --no_cds              Flag. If set, not CDS information will be printed out
                        in the GFF output files. (default: None)
  --source SOURCE       Source field to use for the output files. (default:
                        None)
  --flank FLANK         Flanking distance (in bps) to group non-overlapping
                        transcripts into a single superlocus. Default:
                        determined by the configuration file. (default: None)
  --purge               Flag. If set, the pipeline will suppress any loci
                        whose transcripts do not pass the requirements set in
                        the JSON file. (default: False)
  -shm, --shared-memory
                        Flag. If set, the DB will be copied into memory.
                        (default: False)
  -shmdb SHM_DB, --shared-memory-db SHM_DB
                        Name of the shared memory DB. WARNING: if set, the DB
                        copy will be persistently copied into memory, so that
                        multiple pickers can share. (default: None)
  --preload             Flag. If set, the Mikado DB will be pre-loaded into
                        memory for faster access. WARNING: this option will
                        increase memory usage and the preloading might be
                        quite slow. (default: False)
  -db SQLITE_DB, --sqlite-db SQLITE_DB
                        Location of an SQLite database to overwrite what is
                        specified in the configuration file. (default: None)
  -od OUTPUT_DIR, --output-dir OUTPUT_DIR
                        Output directory. Default: current working directory
                        (default: None)
  --single              Flag. If set, Creator will be launched with a single
                        process. Useful for debugging purposes only. (default:
                        False)
  --mode {nosplit,stringent,lenient,permissive,split}
                        Mode in which Mikado will treat transcripts with
                        multiple ORFs. - nosplit: keep the transcripts whole.
                        - stringent: split multi-orf transcripts if two
                        consecutive ORFs have both BLAST hits and none of
                        those hits is against the same target. - lenient:
                        split multi-orf transcripts as in stringent, and
                        additionally, also when either of the ORFs lacks a
                        BLAST hit (but not both). - permissive: like lenient,
                        but also split when both ORFs lack BLAST hits - split:
                        split multi-orf transcripts regardless of what BLAST
                        data is available. (default: None)

Log options:
  -l LOG, --log LOG     File to write the log to. Default: decided by the
                        configuration file. (default: None)
  -v, --verbose         Flag. If set, the debug mode will be activated.
                        (default: False)
  -nv, --noverbose      Flag. If set, the log will report only errors and
                        critical events. (default: False)
  -lv {DEBUG,INFO,WARN,ERROR,CRITICAL}, --log-level {DEBUG,INFO,WARN,ERROR,CRITICAL}
                        Logging level. Default: retrieved by the configuration
                        file. (default: None)
</pre></div>
</div>
</div>
<div class="section" id="technical-details">
<h2>5.4.2. Technical details<a class="headerlink" href="#technical-details" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">mikado</span> <span class="pre">pick</span></code> uses a divide-et-impera algorithm to find and analyse loci separately. As the data to be integrated with the transcripts is stored on the database rather than be calculated on the fly, rerunning <code class="docutils literal"><span class="pre">pick</span></code> with different options takes little time and resources.
To keep the data sorted, Mikado will write out temporary files during the operation and merge them at the end of the run (see function <code class="docutils literal"><span class="pre">merge_loci_gff</span></code> in <a class="reference internal" href="../Library/picking.html#sub-picking-loci"><span class="std std-ref">the picking module</span></a>.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/mikado-logo.jpg" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">5.4. Mikado pick</a><ul>
<li><a class="reference internal" href="#usage">5.4.1. Usage</a></li>
<li><a class="reference internal" href="#technical-details">5.4.2. Technical details</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Serialise.html"
                        title="previous chapter">5.3. Mikado serialise</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Compare.html"
                        title="next chapter">6.1. Compare</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/Usage_dir/Pick.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Compare.html" title="6.1. Compare"
             >next</a> |</li>
        <li class="right" >
          <a href="Serialise.html" title="5.3. Mikado serialise"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Mikado 1.0.0b4 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../Usage.html" >4. Usage</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Venturini Luca, Caim Shabhonam, Mapleson Daniel, Kaithakottil Gemy George, Swarbreck David.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.5.
    </div>
  </body>
</html>