<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>5.3. Mikado serialise &mdash; Mikado 1.0.0b4 documentation</title>
    
    <link rel="stylesheet" href="../_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../_static/theme_overrides.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.0.0b4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Mikado 1.0.0b4 documentation" href="../index.html" />
    <link rel="up" title="4. Usage" href="../Usage.html" />
    <link rel="next" title="5.4. Mikado pick" href="Pick.html" />
    <link rel="prev" title="5.2. Mikado prepare" href="Prepare.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Pick.html" title="5.4. Mikado pick"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="Prepare.html" title="5.2. Mikado prepare"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Mikado 1.0.0b4 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../Usage.html" accesskey="U">4. Usage</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="mikado-serialise">
<span id="serialise"></span><h1>5.3. Mikado serialise<a class="headerlink" href="#mikado-serialise" title="Permalink to this headline">¶</a></h1>
<p>Mikado integrates data from multiple sources to select the best transcripts. During this step, these sources are brought together inside a common database, simplifying the process of retrieving them at runtime. Currently, Mikado integrates three different types of data:</p>
<div class="sidebar" id="bed12-sidebar">
<p class="first sidebar-title"><a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED12</a> files</p>
<p class="last">During serialisation, Mikado interprets BED12 files as having the relevant information in the <strong>thick-start</strong>, <strong>thick-end</strong> fields, ie columns 7 and 8. For ORF files, this means that the CDS should start at the 7th column and end at the 8th (stop-codon exclusive, like in <a class="reference external" href="http://transdecoder.github.io/">TransDecoder</a>) - or viceversa for ORFs on the negative strand. For junction files, it means that the reliable junction, ie the reliable <em>intron</em>, starts at the thick-start position and ends at the thick-end position. This format requirement is the opposite of what happens for example in the junctions produced by <a class="reference external" href="http://ccb.jhu.edu/software/tophat/index.shtml">TopHat</a>, where the BED12 file lists the two <em>exonic</em> fragments, rather than the intron. <a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> instead provides a tabular non-standard file indicating the most reliable junctions in the alignment. <a class="reference external" href="https://github.com/maplesond/portcullis">Portcullis</a> provides utilities to convert such files into the format used by Mikado, and to merge together multiple junction files. However, we recommend running Portcullis directly on the alignments, rather than using the junctions indicated by the programs themselves.</p>
</div>
<ol class="arabic simple">
<li>reliable junctions, as detected by <a class="reference external" href="https://github.com/maplesond/portcullis">Portcullis</a>, in <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED12</a> format (</li>
<li>ORF data, currently identified using <a class="reference external" href="http://transdecoder.github.io/">TransDecoder</a>, but any program capable of generating it in <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED12</a> format is suitable.</li>
<li>BLASTX <a class="reference internal" href="../References.html#blastplus" id="id1">[Blastplus]</a> data in XML format</li>
</ol>
<p>After serialisation in the database, these sources will be available to use for any subsequent Mikado pick run. Having the data present in the database allows to run Mikado with multiple configurations and little overhead in terms of pre-loading data; this feature is taken directly advtange of in <a class="reference internal" href="Daijin.html#daijin"><span class="std std-ref">The Daijin pipeline for driving Mikado</span></a>, where it is possible to run Mikado using multiple modes.</p>
<p>Mikado serialise can use three different SQL databases as backends - SQLite, MySQL and PostgreSQL - thanks to <a class="reference external" href="http://www.sqlalchemy.org/">SQLAlchemy</a>.
This step, together with the creation of the TransDecoder and BLAST data, is the most time consuming of the pipeline. In particular, although Mikado serialise will try to analyse the XML data in a parallelised fashion if so instructed, the insertion of the data in the database will still happen in a single thread and will therefore be of limited speed. If using SQLite as database (the default option), it is possible to decrease the runtime by modifying the &#8220;max_objects&#8221; parameters, at the cost however of increased RAM usage.</p>
<div class="section" id="transdecoder-orfs">
<h2>5.3.1. Transdecoder ORFs<a class="headerlink" href="#transdecoder-orfs" title="Permalink to this headline">¶</a></h2>
<p>When Mikado analyses ORFs produced by <a class="reference external" href="http://transdecoder.github.io/">TransDecoder</a> or equivalent program, it performs additionally the following checks:</p>
<ol class="arabic simple">
<li>Check the congruence between the length of the transcript in the BED12 file and that found in the FASTA file</li>
<li>Check that the ORF does not contain internal stop codons</li>
<li>Check that the CDS length is valid, ie a multiple of 3, if the ORF is complete</li>
<li>Optionally, if the ORF is open on the 5&#8217; side, Mikado can try to find an internal start codon. See :<cite>this section &lt;max-regression&gt;</cite> for details.</li>
</ol>
</div>
<div class="section" id="usage">
<h2>5.3.2. Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h2>
<p><code class="docutils literal"><span class="pre">mikado</span> <span class="pre">serialise</span></code> allows to override some of the parameters present in the configuration file through command line options, eg the input files. Notwithstanding, in the interest of reproducibility we advise to configure everything through the configuration file and supply it to Mikado prepare without further modifications.</p>
<p>Available parameters:</p>
<ul>
<li><dl class="first docutils">
<dt>Parameters related to performance:</dt>
<dd><ul class="first last simple">
<li><em>start-method</em>: one of fork, spawn, forkserver. It determines the multiprocessing start method. By default, Mikado will use the default for the system (fork on UNIX, spawn on Windows).</li>
<li><em>procs</em>: Number of processors to use.</li>
<li><em>single-thread</em>: flag. If set, Mikado will disable all multithreading.</li>
<li><em>max_objects</em>: Maximum number of objects to keep in memory before committing to the database. See <a class="reference internal" href="Configure.html#max-objects"><span class="std std-ref">this section of the configuration</span></a> for details.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Basic input data and settings:</dt>
<dd><ul class="first last simple">
<li><em>output-dir</em>: directory where the SQLite database and the log will be written to.</li>
<li><em>transcripts</em>: these are the input transcripts that are present on the GTF file considered by Mikado. Normally this should be the output of Mikado prepare.</li>
<li><em>genome_fai</em>: FAIDX file of the genome FASTA. If not given, serialise will derive it from the &#8220;reference: genome&#8221; field of the configuration.</li>
<li><em>force</em>: flag. If set, and the database is already present, it will be truncated rather than updated.</li>
<li><strong>json-conf</strong>: this is the configuration file created with <a class="reference internal" href="Configure.html#configure"><span class="std std-ref">Mikado configure</span></a>.</li>
<li><em>db</em>: if the database is specified on the command line, <code class="docutils literal"><span class="pre">mikado</span> <span class="pre">serialise</span></code> will interpret it as a <strong>SQLite</strong> database. This will overwrite any setting present in the configuration file.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Parameters related to logging:</dt>
<dd><ul class="first last simple">
<li><em>log</em>: log file. It defaults to <code class="docutils literal"><span class="pre">serialise.log</span></code>.</li>
<li><em>log_level</em>: verbosity of the logging. Please be advised that excessive verbosity can negatively impact the performance of the program - the debug mode is extremely verbose.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Parameters related to reliable junctions:</dt>
<dd><ul class="first last simple">
<li><em>junctions</em>: a <a class="reference external" href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1">BED12</a> file of reliable junctions. This can be obtained using <a class="reference external" href="https://github.com/maplesond/portcullis">Portcullis</a>. Please see the relative <a class="reference internal" href="#bed12-sidebar"><span class="std std-ref">sidebar</span></a>.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Parameters related to the treatment of ORF data:</dt>
<dd><ul class="first last simple">
<li><em>orfs</em>: ORF BED12 files, separated by comma.</li>
<li><em>max-regression</em>: A percentage, expressed as a number between 0 and 1, which indicates how far can Mikado regress along the ORF to find a valid start codon. See the <a class="reference internal" href="Configure.html#max-regression"><span class="std std-ref">relative section in the configuration</span></a> for details.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>Parameters related to BLAST data:</dt>
<dd><ul class="first last">
<li><p class="first"><em>blast_targets</em>: BLAST FASTA database.</p>
</li>
<li><p class="first"><em>discard-definition</em>: Flag. Depending on how the database has been created, sometimes BLAST will substitute the ID of the sequence with &#8220;lcl|&#8221; ids. Mikado circumvents this by looking for the definition field in the XML file. Using this flag will disable this behaviour and force Mikado to use the ID - with the potential of having a mismatch between the sequences in the BLAST DB and the sequences in the BLAST files.</p>
</li>
<li><dl class="first docutils">
<dt><em>xml</em>: BLAST files to parse. This can be one of the following:</dt>
<dd><ul class="first last simple">
<li>A list of XML BLAST files, optionally compressed with GZip or BZip2, comma separated (suffix .xml)</li>
<li>A list of ASN BLAST files, optionally compressed with GZip or BZip2, comma separated (suffix .asn)</li>
<li>A list of folders, comma separated, where it is possible to find files of the former 2 types</li>
<li>A mixture of the three above types.</li>
</ul>
</dd>
</dl>
</li>
<li><p class="first"><em>max-target-seqs</em>: maximum number of BLAST targets that can be loaded per sequence, for each BLAST alignment. Please note that if you align against multiple databases, this threshold will be applied once per file.</p>
</li>
</ul>
</dd>
</dl>
</li>
</ul>
<div class="admonition hint">
<p class="first admonition-title">Hint</p>
<p class="last">Mikado will parallelise only the reading of multiple XML files. As such, this part of the pipeline is less performing than the other steps.</p>
</div>
<div class="admonition warning">
<p class="first admonition-title">Warning</p>
<p class="last">It is advised to set this parameter to <em>spawn</em> even on UNIX. See <a class="reference internal" href="Configure.html#scheduler-multiprocessing"><span class="std std-ref">the dedicated sidebar for details</span></a>.</p>
</div>
<p>Usage:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ mikado serialise --help
usage: Mikado serialise [-h] [--start-method {fork,spawn,forkserver}]
                        [--orfs ORFS] [--transcripts TRANSCRIPTS]
                        [-mr MAX_REGRESSION]
                        [--max_target_seqs MAX_TARGET_SEQS]
                        [--blast_targets BLAST_TARGETS] [--discard-definition]
                        [--xml XML] [-p PROCS] [--single-thread]
                        [--genome_fai GENOME_FAI] [--junctions JUNCTIONS]
                        [-mo MAX_OBJECTS] [-f] --json-conf JSON_CONF
                        [-l [LOG]] [-od OUTPUT_DIR]
                        [-lv {DEBUG,INFO,WARN,ERROR}]
                        [db]

Mikado serialise creates the database used by the pick program. It handles
Junction and ORF BED12 files as well as BLAST XML results.

optional arguments:
  -h, --help            show this help message and exit
  --start-method {fork,spawn,forkserver}
                        Multiprocessing start method.
  -od OUTPUT_DIR, --output-dir OUTPUT_DIR
                        Output directory. Default: current working directory

  --orfs ORFS           ORF BED file(s), separated by commas
  --transcripts TRANSCRIPTS
                        Transcript FASTA file(s) used for ORF calling and
                        BLAST queries, separated by commas. If multiple files
                        are given, they must be in the same order of the ORF
                        files. E.g. valid command lines are:
                        --transcript_fasta all_seqs1.fasta --orfs all_orfs.bed
                        --transcript_fasta seq1.fasta,seq2.fasta --orfs
                        orfs1.bed,orf2.bed --transcript_fasta all_seqs.fasta
                        --orfs orfs1.bed,orf2.bed These are invalid instead: #
                        Inverted order --transcript_fasta
                        seq1.fasta,seq2.fasta --orfs orfs2.bed,orf1.bed #Two
                        transcript files, one ORF file --transcript_fasta
                        seq1.fasta,seq2.fasta --orfs all_orfs.bed
  -mr MAX_REGRESSION, --max-regression MAX_REGRESSION
                        &quot;Amount of sequence in the ORF (in %) to backtrack in
                        order to find a valid START codon, if one is absent.
                        Default: None

  --max_target_seqs MAX_TARGET_SEQS
                        Maximum number of target sequences.
  --blast_targets BLAST_TARGETS
                        Target sequences
  --discard-definition  Flag. If set, the sequences IDs instead of their
                        definition will be used for serialisation.
  --xml XML             XML file(s) to parse. They can be provided in three
                        ways: - a comma-separated list - as a base folder -
                        using bash-like name expansion (*,?, etc.). In this
                        case, you have to enclose the filename pattern in
                        double quotes. Multiple folders/file patterns can be
                        given, separated by a comma.
  -p PROCS, --procs PROCS
                        Number of threads to use for analysing the BLAST
                        files. This number should not be higher than the total
                        number of XML files.
  --single-thread       Force serialise to run with a single thread,
                        irrespective of other configuration options.

  --genome_fai GENOME_FAI
  --junctions JUNCTIONS

  -mo MAX_OBJECTS, --max-objects MAX_OBJECTS
                        Maximum number of objects to cache in memory before
                        committing to the database. Default: 100,000 i.e.
                        approximately 450MB RAM usage for Drosophila.
  -f, --force           Flag. If set, an existing databse will be deleted
                        (sqlite) or dropped (MySQL/PostGreSQL) before
                        beginning the serialisation.
  --json-conf JSON_CONF
  -l [LOG], --log [LOG]
                        Optional log file. Default: stderr
  -lv {DEBUG,INFO,WARN,ERROR}, --log_level {DEBUG,INFO,WARN,ERROR}
                        Log level. Default: INFO
  db                    Optional output database. Default: derived from
                        json_conf
</pre></div>
</div>
</div>
<div class="section" id="technical-details">
<h2>5.3.3. Technical details<a class="headerlink" href="#technical-details" title="Permalink to this headline">¶</a></h2>
<p>The schema of the database is quite simple, as it is composed only of 7 discrete tables in two groups. The first group, <em>chrom</em> and <em>junctions</em>, serialises the information pertaining to the reliable junctions - ie information which is not relative to the transcripts but rather to their genomic locations.
The second group serialises the data regarding ORFs and BLAST files. The need of using a database is mainly driven by the latter, as querying a relational database is faster than retrieving the information from the XML files themselves at runtime.</p>
<div class="topic">
<p class="topic-title first">Database schema used by Mikado.</p>
<div class="figure align-center">
<a class="reference internal image-reference" href="../_images/database_schema.png"><img alt="../_images/database_schema.png" src="../_images/database_schema.png" style="width: 323.5px; height: 567.5px;" /></a>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="../index.html">
              <img class="logo" src="../_static/mikado-logo.jpg" alt="Logo"/>
            </a></p>
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">5.3. Mikado serialise</a><ul>
<li><a class="reference internal" href="#transdecoder-orfs">5.3.1. Transdecoder ORFs</a></li>
<li><a class="reference internal" href="#usage">5.3.2. Usage</a></li>
<li><a class="reference internal" href="#technical-details">5.3.3. Technical details</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="Prepare.html"
                        title="previous chapter">5.2. Mikado prepare</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Pick.html"
                        title="next chapter">5.4. Mikado pick</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/Usage_dir/Serialise.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="../py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Pick.html" title="5.4. Mikado pick"
             >next</a> |</li>
        <li class="right" >
          <a href="Prepare.html" title="5.2. Mikado prepare"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">Mikado 1.0.0b4 documentation</a> &raquo;</li>
          <li class="nav-item nav-item-1"><a href="../Usage.html" >4. Usage</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Venturini Luca, Caim Shabhonam, Mapleson Daniel, Kaithakottil Gemy George, Swarbreck David.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.5.
    </div>
  </body>
</html>