<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Mikado.subprograms.util.grep &mdash; Mikado 19 documentation</title>
    
    <link rel="stylesheet" href="../../../../_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="../../../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../../../',
        VERSION:     '19',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../../../_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Mikado 19 documentation" href="../../../../index.html" />
    <link rel="up" title="Mikado.subprograms.util" href="../util.html" />
   
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body role="document">  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <h1>Source code for Mikado.subprograms.util.grep</h1><div class="highlight"><pre>
<span></span><span class="ch">#!/usr/bin/env python3</span>
<span class="c1"># coding: utf-8</span>

<span class="sd">&quot;&quot;&quot;Script to parse and retrieve given features from a GTF/GFF file.&quot;&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">sys</span>
<span class="kn">import</span> <span class="nn">argparse</span>
<span class="kn">import</span> <span class="nn">functools</span>
<span class="kn">from</span> <span class="nn">...parsers</span> <span class="kn">import</span> <span class="n">GFF</span><span class="p">,</span> <span class="n">GTF</span>
<span class="kn">from</span> <span class="nn">...subprograms</span> <span class="kn">import</span> <span class="n">to_gff</span>

<span class="n">__author__</span> <span class="o">=</span> <span class="s1">&#39;Luca Venturini&#39;</span>


<div class="viewcode-block" id="print_gff_gene"><a class="viewcode-back" href="../../../../Mikado.subprograms.util.html#Mikado.subprograms.util.grep.print_gff_gene">[docs]</a><span class="k">def</span> <span class="nf">print_gff_gene</span><span class="p">(</span><span class="n">curr_gene</span><span class="p">,</span> <span class="n">curr_transcripts</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    :param curr_gene: gene record</span>
<span class="sd">    :param curr_transcripts: dictionary of transcripts</span>
<span class="sd">    :param args: argparse</span>
<span class="sd">    :return: None</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">if</span> <span class="n">curr_gene</span> <span class="ow">is</span> <span class="ow">not</span> <span class="bp">None</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">curr_transcripts</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">starts</span><span class="p">,</span> <span class="n">ends</span> <span class="o">=</span> <span class="p">[],</span> <span class="p">[]</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">tid</span> <span class="ow">in</span> <span class="n">curr_transcripts</span><span class="p">:</span>
            <span class="n">lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">curr_transcripts</span><span class="p">[</span><span class="n">tid</span><span class="p">][</span><span class="mi">0</span><span class="p">]))</span>
            <span class="n">starts</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_transcripts</span><span class="p">[</span><span class="n">tid</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">start</span><span class="p">)</span>
            <span class="n">ends</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">curr_transcripts</span><span class="p">[</span><span class="n">tid</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">end</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">rec</span> <span class="ow">in</span> <span class="n">curr_transcripts</span><span class="p">[</span><span class="n">tid</span><span class="p">][</span><span class="mi">1</span><span class="p">:]:</span>
                <span class="n">rec</span><span class="o">.</span><span class="n">parent</span> <span class="o">=</span> <span class="p">[</span><span class="n">tid</span><span class="p">]</span>
                <span class="n">lines</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="n">rec</span><span class="p">))</span>
        <span class="n">curr_gene</span><span class="o">.</span><span class="n">start</span> <span class="o">=</span> <span class="nb">min</span><span class="p">(</span><span class="n">starts</span><span class="p">)</span>
        <span class="n">curr_gene</span><span class="o">.</span><span class="n">end</span> <span class="o">=</span> <span class="nb">max</span><span class="p">(</span><span class="n">ends</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="n">curr_gene</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="o">*</span><span class="n">lines</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="s2">&quot;###&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span></div>


<div class="viewcode-block" id="verify_storability"><a class="viewcode-back" href="../../../../Mikado.subprograms.util.html#Mikado.subprograms.util.grep.verify_storability">[docs]</a><span class="k">def</span> <span class="nf">verify_storability</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">mrna_ids</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function verifies whether a GFF transcript has to be kept</span>
<span class="sd">    or not.</span>
<span class="sd">    :param record: the record to be evaluated</span>
<span class="sd">    :param gene_ids: a set of gene IDs</span>
<span class="sd">    :param mrna_ids: a set of mRNA IDs</span>
<span class="sd">    :param args: the namespace args</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">bool_flag</span> <span class="o">=</span> <span class="bp">False</span>

    <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">reverse</span> <span class="ow">is</span> <span class="bp">False</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">in</span> <span class="n">mrna_ids</span><span class="p">:</span>
            <span class="n">bool_flag</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">genes</span> <span class="ow">is</span> <span class="bp">True</span> <span class="ow">and</span> <span class="nb">any</span><span class="p">([</span><span class="n">p</span> <span class="ow">in</span> <span class="n">gene_ids</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">parent</span><span class="p">]):</span>
            <span class="n">bool_flag</span> <span class="o">=</span> <span class="bp">True</span>
            <span class="n">mrna_ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">reverse</span> <span class="ow">is</span> <span class="bp">True</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">genes</span> <span class="ow">is</span> <span class="bp">False</span> <span class="ow">and</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mrna_ids</span><span class="p">:</span>
            <span class="n">bool_flag</span> <span class="o">=</span> <span class="bp">True</span>
        <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">genes</span> <span class="ow">is</span> <span class="bp">True</span> <span class="ow">and</span> <span class="ow">not</span> <span class="nb">any</span><span class="p">([</span><span class="n">p</span> <span class="ow">in</span> <span class="n">gene_ids</span> <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">record</span><span class="o">.</span><span class="n">parent</span><span class="p">]):</span>
            <span class="n">bool_flag</span> <span class="o">=</span> <span class="bp">True</span>
            <span class="n">mrna_ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">bool_flag</span><span class="p">,</span> <span class="n">mrna_ids</span></div>


<div class="viewcode-block" id="grep_gff"><a class="viewcode-back" href="../../../../Mikado.subprograms.util.html#Mikado.subprograms.util.grep.grep_gff">[docs]</a><span class="k">def</span> <span class="nf">grep_gff</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">mrna_ids</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Grep-like main function for *GFF3* files.</span>
<span class="sd">    :param args: argparse Namespace with the options.</span>

<span class="sd">    :param gene_ids: set of gene IDs to retain/discard.</span>
<span class="sd">    :type gene_ids: set</span>

<span class="sd">    :param mrna_ids: set of mRNA IDs to retain/discard.</span>
<span class="sd">    :type mrna_ids: set</span>

<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">curr_gene</span> <span class="o">=</span> <span class="bp">None</span>
    <span class="n">curr_transcripts</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="k">print</span><span class="p">(</span><span class="s2">&quot;##gff-version 3&quot;</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>

    <span class="n">evaluator</span> <span class="o">=</span> <span class="n">functools</span><span class="o">.</span><span class="n">partial</span><span class="p">(</span><span class="n">verify_storability</span><span class="p">,</span>
                                  <span class="o">**</span><span class="p">{</span><span class="s2">&quot;gene_ids&quot;</span><span class="p">:</span> <span class="n">gene_ids</span><span class="p">,</span>
                                     <span class="s2">&quot;args&quot;</span><span class="p">:</span> <span class="n">args</span><span class="p">})</span>

    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">gff</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">is_transcript</span> <span class="ow">is</span> <span class="bp">True</span><span class="p">:</span>
            <span class="n">evaluated</span><span class="p">,</span> <span class="n">mrna_ids</span> <span class="o">=</span> <span class="n">evaluator</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="n">mrna_ids</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">evaluated</span> <span class="ow">is</span> <span class="bp">True</span><span class="p">:</span>
                <span class="n">curr_transcripts</span><span class="p">[</span><span class="n">record</span><span class="o">.</span><span class="n">id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">record</span><span class="p">]</span>
        <span class="c1"># Records with &quot;derived from&quot; are typically things like &quot;protein&quot; from TAIR</span>
        <span class="k">elif</span> <span class="n">record</span><span class="o">.</span><span class="n">is_exon</span> <span class="ow">is</span> <span class="bp">True</span> <span class="ow">or</span> <span class="n">record</span><span class="o">.</span><span class="n">is_derived</span> <span class="ow">is</span> <span class="bp">True</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">is_derived</span> <span class="ow">is</span> <span class="bp">True</span><span class="p">:</span>
                <span class="n">parent_store</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">derived_from</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">parent_store</span> <span class="o">=</span> <span class="n">record</span><span class="o">.</span><span class="n">parent</span>
            <span class="k">for</span> <span class="n">parent</span> <span class="ow">in</span> <span class="n">parent_store</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">parent</span> <span class="ow">in</span> <span class="n">mrna_ids</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">reverse</span> <span class="ow">is</span> <span class="bp">False</span><span class="p">:</span>
                    <span class="n">curr_transcripts</span><span class="p">[</span><span class="n">parent</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">parent</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mrna_ids</span> <span class="ow">and</span> <span class="n">args</span><span class="o">.</span><span class="n">reverse</span> <span class="ow">is</span> <span class="bp">True</span><span class="p">:</span>
                    <span class="n">curr_transcripts</span><span class="p">[</span><span class="n">parent</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">print_gff_gene</span><span class="p">(</span><span class="n">curr_gene</span><span class="p">,</span> <span class="n">curr_transcripts</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span>
            <span class="n">curr_gene</span> <span class="o">=</span> <span class="bp">None</span>
            <span class="n">curr_transcripts</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">()</span>

            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">id</span> <span class="ow">is</span> <span class="bp">None</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="n">curr_gene</span> <span class="o">=</span> <span class="n">record</span>

    <span class="n">print_gff_gene</span><span class="p">(</span><span class="n">curr_gene</span><span class="p">,</span> <span class="n">curr_transcripts</span><span class="p">,</span> <span class="n">args</span><span class="p">)</span></div>


<div class="viewcode-block" id="grep_gtf"><a class="viewcode-back" href="../../../../Mikado.subprograms.util.html#Mikado.subprograms.util.grep.grep_gtf">[docs]</a><span class="k">def</span> <span class="nf">grep_gtf</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">mrna_ids</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Grep-like main function for *GTF* files.</span>
<span class="sd">    :param args:</span>
<span class="sd">    :return:</span>

<span class="sd">    :param gene_ids: set of gene IDs to retain/discard.</span>
<span class="sd">    :type gene_ids: set</span>

<span class="sd">    :param mrna_ids: set of mRNA IDs to retain/discard.</span>
<span class="sd">    :type mrna_ids: set</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">gff</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">record</span><span class="p">:</span>
            <span class="k">continue</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">genes</span> <span class="ow">is</span> <span class="bp">False</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">transcript</span> <span class="ow">in</span> <span class="n">mrna_ids</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">reverse</span><span class="p">:</span>
                    <span class="k">print</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">reverse</span><span class="p">:</span>
                    <span class="k">print</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_ids</span><span class="p">:</span>
                <span class="k">if</span> <span class="ow">not</span> <span class="n">args</span><span class="o">.</span><span class="n">reverse</span><span class="p">:</span>
                    <span class="k">print</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span>
                <span class="k">elif</span> <span class="n">args</span><span class="o">.</span><span class="n">reverse</span><span class="p">:</span>
                    <span class="k">print</span><span class="p">(</span><span class="n">record</span><span class="p">,</span> <span class="nb">file</span><span class="o">=</span><span class="n">args</span><span class="o">.</span><span class="n">out</span><span class="p">)</span></div>


<div class="viewcode-block" id="launch"><a class="viewcode-back" href="../../../../Mikado.subprograms.util.html#Mikado.subprograms.util.grep.launch">[docs]</a><span class="k">def</span> <span class="nf">launch</span><span class="p">(</span><span class="n">args</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Function which chooses whether to use the grep_gtf</span>
<span class="sd">    or the grep_gff function for the analysis.</span>
<span class="sd">    :param args:</span>
<span class="sd">    :return:</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">gene_ids</span><span class="p">,</span> <span class="n">mrna_ids</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(),</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">args</span><span class="o">.</span><span class="n">ids</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">args</span><span class="o">.</span><span class="n">genes</span> <span class="ow">is</span> <span class="bp">False</span><span class="p">:</span>
            <span class="n">mrna_id</span><span class="p">,</span> <span class="n">gene_id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">()[:</span><span class="mi">2</span><span class="p">]</span>
            <span class="n">mrna_ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">mrna_id</span><span class="p">)</span>
            <span class="n">gene_ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">gene_id</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">rstrip</span><span class="p">()</span>
            <span class="n">gene_ids</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>

    <span class="k">if</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gff</span><span class="p">,</span> <span class="n">GTF</span><span class="o">.</span><span class="n">GTF</span><span class="p">):</span>
        <span class="n">grep_gtf</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">mrna_ids</span><span class="p">)</span>
    <span class="k">elif</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gff</span><span class="p">,</span> <span class="n">GFF</span><span class="o">.</span><span class="n">GFF3</span><span class="p">):</span>
        <span class="n">grep_gff</span><span class="p">(</span><span class="n">args</span><span class="p">,</span> <span class="n">gene_ids</span><span class="p">,</span> <span class="n">mrna_ids</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">TypeError</span><span class="p">(</span><span class="nb">type</span><span class="p">(</span><span class="n">args</span><span class="o">.</span><span class="n">gff</span><span class="p">))</span></div>


<div class="viewcode-block" id="grep_parser"><a class="viewcode-back" href="../../../../Mikado.subprograms.util.html#Mikado.subprograms.util.grep.grep_parser">[docs]</a><span class="k">def</span> <span class="nf">grep_parser</span><span class="p">():</span>
    <span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Simple command line parser for the utility.</span>
<span class="sd">    :return: args</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">parser</span> <span class="o">=</span> <span class="n">argparse</span><span class="o">.</span><span class="n">ArgumentParser</span><span class="p">(</span>
        <span class="s1">&#39;Script to parse and retrieve given features from a GFF file.&#39;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;-v&#39;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s1">&#39;store_true&#39;</span><span class="p">,</span> <span class="n">dest</span><span class="o">=</span><span class="s1">&#39;reverse&#39;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Exclude from the gff all the records in the id file.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s2">&quot;--genes&quot;</span><span class="p">,</span> <span class="n">action</span><span class="o">=</span><span class="s2">&quot;store_true&quot;</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;&quot;&quot;Flag. If set, the program expects as ids</span>
<span class="s2">                        only a list of genes, and will exclude/include all the transcripts</span>
<span class="s2">                        children of the selected genes.&quot;&quot;&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;ids&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">FileType</span><span class="p">(),</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;ID file (format: mrna_id, gene_id - tab separated)&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;gff&#39;</span><span class="p">,</span> <span class="nb">type</span><span class="o">=</span><span class="n">to_gff</span><span class="p">,</span> <span class="n">help</span><span class="o">=</span><span class="s2">&quot;The GFF file to parse.&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">add_argument</span><span class="p">(</span><span class="s1">&#39;out&#39;</span><span class="p">,</span> <span class="n">nargs</span><span class="o">=</span><span class="s1">&#39;?&#39;</span><span class="p">,</span>
                        <span class="nb">type</span><span class="o">=</span><span class="n">argparse</span><span class="o">.</span><span class="n">FileType</span><span class="p">(</span><span class="s1">&#39;w&#39;</span><span class="p">),</span> <span class="n">default</span><span class="o">=</span><span class="n">sys</span><span class="o">.</span><span class="n">stdout</span><span class="p">,</span>
                        <span class="n">help</span><span class="o">=</span><span class="s2">&quot;Optional output file&quot;</span><span class="p">)</span>
    <span class="n">parser</span><span class="o">.</span><span class="n">set_defaults</span><span class="p">(</span><span class="n">func</span><span class="o">=</span><span class="n">launch</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">parser</span></div>
</pre></div>

          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="../../../../index.html">Documentation overview</a><ul>
  <li><a href="../../../index.html">Module code</a><ul>
  <li><a href="../util.html">Mikado.subprograms.util</a><ul>
  </ul></li>
  </ul></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../../../../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2016, Luca Venturini, Shabhonam Caim.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.7</a>
      
    </div>

    

    
  </body>
</html>