

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>1. Introduction &mdash; Mikado 1.0.0b4 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
  

  
    <link rel="top" title="Mikado 1.0.0b4 documentation" href="index.html"/>
        <link rel="next" title="2. Installation" href="Installation.html"/>
        <link rel="prev" title="Mikado: pick your transcript" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Mikado
          

          
          </a>

          
            
            
              <div class="version">
                1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">1. Introduction</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#metrics">1.1. Metrics</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="Installation.html">2. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial.html">3. Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="Usage.html">4. Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="Usage.html#mikado-pipeline-stages">5. Mikado pipeline stages</a></li>
<li class="toctree-l1"><a class="reference internal" href="Usage.html#mikado-utilities">6. Mikado utilities</a></li>
<li class="toctree-l1"><a class="reference internal" href="References.html">7. References</a></li>
<li class="toctree-l1"><a class="reference internal" href="Library/index.html">8. Library documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Mikado</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>1. Introduction</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/Introduction.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <span class="target" id="introduction"></span><div class="section" id="id1">
<h1>1. Introduction<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://www.python.org/"><img alt="python_badge" src="https://img.shields.io/pypi/pyversions/snakemake.svg?style=flat-square" /></a> <a class="reference external" href="http://snakemake.bitbucket.org"><img alt="snake_badge" src="https://img.shields.io/badge/snakemake-≥3.5.2-brightgreen.svg?style=flat-square)]" /></a></p>
<p>Introduction test</p>
<div class="section" id="metrics">
<span id="id2"></span><h2>1.1. Metrics<a class="headerlink" href="#metrics" title="Permalink to this headline">¶</a></h2>
<p>These are all the metrics available to quantify transcripts. The documentation for this section has been generated using the <a class="reference internal" href="Usage_dir/Utilities.html#metrics-command"><span class="std std-ref">metrics utility</span></a>.</p>
<p>Metrics belong to one of the following categories:</p>
<ul class="simple">
<li><strong>Descriptive</strong>: these metrics merely provide a description of the transcript (eg its ID) and are not used for scoring.</li>
<li><strong>cDNA</strong>: these metrics refer to basic features of any transcript such as its number of exons, its cDNA length, etc.</li>
<li><strong>Intron</strong>: these metrics refer to features related to the number of introns and their lengths.</li>
<li><strong>CDS</strong>: these metrics refer to features related to the CDS assigned to the transcript.</li>
<li><strong>UTR</strong>: these metrics refer to features related to the UTR of the transcript. In the case in which a transcript has been assigned multiple ORFs, unless otherwise stated the UTR metrics will be derived only considering the <em>selected</em> ORF, not the combination of all of them.</li>
<li><strong>Locus</strong>: these metrics refer to features of the transcript in relationship to all other transcripts in its locus, eg how many of the introns present in the locus are present in the transcript. These metrics are calculated by Mikado during the picking phase, and as such their value can vary during the different stages as the transcripts are shifted to different groups.</li>
<li><strong>External</strong>: these metrics are derived from accessory data that is recovered for the transcript during the run time. Examples include data regarding the number of introns confirmed by external programs such as PortCullis, or the BLAST score of the best hits.</li>
</ul>
<table border="1" class="docutils">
<colgroup>
<col width="35%" />
<col width="43%" />
<col width="10%" />
<col width="12%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Metric name</th>
<th class="head">Description</th>
<th class="head">Data type</th>
<th class="head">Category</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td><em>tid</em></td>
<td>Name of the transcript. Not used for scoring.</td>
<td>String</td>
<td><strong>Descriptive</strong></td>
</tr>
<tr class="row-odd"><td><em>parent</em></td>
<td>Name of the transcript parent. Not used for scoring.</td>
<td>String</td>
<td><strong>Descriptive</strong></td>
</tr>
<tr class="row-even"><td><em>score</em></td>
<td>Final score of the transcript.</td>
<td>Float</td>
<td><strong>Descriptive</strong></td>
</tr>
<tr class="row-odd"><td><em>best_bits</em></td>
<td>Best Bit Score associated with the transcript.</td>
<td>Float</td>
<td><strong>External</strong></td>
</tr>
<tr class="row-even"><td><em>blast_score</em></td>
<td>Alias for either <em>best_bits</em> or <em>snowy_blast_score</em>. Set
currently to <em>best_bits</em></td>
<td>Float</td>
<td><strong>External</strong></td>
</tr>
<tr class="row-odd"><td><em>canonical_intron_proportion</em></td>
<td>This metric returns the proportion of canonical introns of
the transcript on its total number of introns.</td>
<td>Float
Float</td>
<td><strong>Intron</strong></td>
</tr>
<tr class="row-even"><td><em>cdna_length</em></td>
<td>This property returns the length of the transcript.</td>
<td>Int</td>
<td><strong>cDNA</strong></td>
</tr>
<tr class="row-odd"><td><em>cds_not_maximal</em></td>
<td>This property returns the length of the CDS excluding
that contained in the selected ORF. If the transcript only
has one ORF, this metric returns a value of 0.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>cds_not_maximal_fraction</em></td>
<td>This property returns the fraction of bases not in the
selected ORF compared to the total number of CDS bases
in the cDNA.</td>
<td>Float</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>combined_cds_fraction</em></td>
<td>This property return the percentage of the CDS part of the
transcript vs. the cDNA length.</td>
<td>Float</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>combined_cds_intron_fraction</em></td>
<td>This property returns the fraction of CDS introns of the
transcript vs. the total number of CDS introns in the
Locus. If the transcript is by itself, it returns 1.</td>
<td>Float</td>
<td><strong>Locus</strong></td>
</tr>
<tr class="row-odd"><td><em>combined_cds_length</em></td>
<td>This property returns the fraction of CDS introns of the
transcript, across all its ORFs.</td>
<td>Float</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>combined_cds_num</em></td>
<td>This property returns the number of non-overlapping CDS
segments in the transcript.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>combined_cds_num_fraction</em></td>
<td>This property returns the fraction of non-overlapping CDS
segments in the transcript vs. the total number of exons.</td>
<td>Float</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>combined_utr_fraction</em></td>
<td>This property returns the fraction of the cDNA which is
not coding according to any ORF. Complement of
<em>combined_cds_fraction</em></td>
<td>Float</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-odd"><td><em>combined_utr_length</em></td>
<td>This property return the length of the UTR part of the
transcript.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-even"><td><em>end_distance_from_junction</em></td>
<td>This metric returns the cDNA distance between the stop
and the last junction of the transcript. In many
eukaryotes, this distance cannot exceed 50-55 bps,
otherwise the transcript becomes a target for NMD. If the
transcript is not coding or there is no junction
downstream of the stop codon, the metric returns 0.
This metric considers the combined CDS end.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>end_distance_from_tes</em></td>
<td>This property returns the distance of the end of the
combined CDS from the transcript end site. If no CDS is
defined, it defaults to 0.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>exon_fraction</em></td>
<td>This property returns the fraction of exons of the
transcript which are contained in the sublocus. If the
transcript is by itself, it returns 1.</td>
<td>Float</td>
<td><strong>Locus</strong></td>
</tr>
<tr class="row-odd"><td><em>exon_num</em></td>
<td>This property returns the number of exons of the
transcript.</td>
<td>Int</td>
<td><strong>cDNA</strong></td>
</tr>
<tr class="row-even"><td><em>five_utr_length</em></td>
<td>Returns the length of the 5&#8217; UTR of the <em>selected</em> ORF.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-odd"><td><em>five_utr_num</em></td>
<td>This property returns the number of 5&#8217; UTR segments for
the selected ORF.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-even"><td><em>five_utr_num_complete</em></td>
<td>This property returns the number of 5&#8217; UTR segments for
the selected ORF, considering only those which are
complete exons.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-odd"><td><em>has_start_codon</em></td>
<td>True if the selected ORF has a start codon, False
otherwise</td>
<td>Bool</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>has_stop_codon</em></td>
<td>True if the selected ORF has a stop codon, False otherwise</td>
<td>Bool</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>highest_cds_exon_number</em></td>
<td>This property returns the maximum number of CDS segments
among the ORFs; this number can refer to an ORF
<em>DIFFERENT</em> from the maximal ORF.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>highest_cds_exons_num</em></td>
<td>Returns the number of CDS segments in the selected ORF
(irrespective of the number of exons involved)</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>intron_fraction</em></td>
<td>This property returns the fraction of introns of the
transcript vs. the total number of introns in the Locus.
If the transcript is by itself, it returns 1.</td>
<td>Float</td>
<td><strong>Locus</strong></td>
</tr>
<tr class="row-even"><td><em>is_complete</em></td>
<td>Boolean. True if the selected ORF has both start and end.</td>
<td>Bool</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>max_intron_length</em></td>
<td>This property returns the greatest intron length for the
transcript.</td>
<td>Int</td>
<td><strong>Intron</strong></td>
</tr>
<tr class="row-even"><td><em>min_intron_length</em></td>
<td>This property returns the smallest intron length for the
transcript.</td>
<td>Int</td>
<td><strong>Intron</strong></td>
</tr>
<tr class="row-odd"><td><em>non_verified_introns_num</em></td>
<td>This metric returns the number of introns of the
transcript which are not validated by external data.</td>
<td>Int</td>
<td><strong>External</strong></td>
</tr>
<tr class="row-even"><td><em>num_introns_greater_than_max</em></td>
<td>This metric returns the number of introns greater than the
maximum acceptable intron size indicated in the
constructor.</td>
<td>Int</td>
<td><strong>Intron</strong></td>
</tr>
<tr class="row-odd"><td><em>num_introns_smaller_than_min</em></td>
<td>This metric returns the number of introns smaller than the
mininum acceptable intron size indicated in the
constructor.</td>
<td>Int</td>
<td><strong>Intron</strong></td>
</tr>
<tr class="row-even"><td><em>number_internal_orfs</em></td>
<td>This property returns the number of ORFs inside a
transcript.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>proportion_verified_introns</em></td>
<td>This metric returns, as a fraction, how many of the
transcript introns are validated by external data.</td>
<td>Float</td>
<td><strong>External</strong></td>
</tr>
<tr class="row-even"><td><em>proportion_verified_introns_inlocus</em></td>
<td>This metric returns, as a fraction, how many of the
verified introns inside the Locus are contained inside the
transcript.</td>
<td>Float</td>
<td><strong>Locus</strong></td>
</tr>
<tr class="row-odd"><td><em>retained_fraction</em></td>
<td>This property returns the fraction of the cDNA which is
contained in retained introns.</td>
<td>Float</td>
<td><strong>Locus</strong></td>
</tr>
<tr class="row-even"><td><em>retained_intron_num</em></td>
<td>This property records the number of introns in the
transcripts which are marked as being retained.</td>
<td>Int</td>
<td><strong>Locus</strong></td>
</tr>
<tr class="row-odd"><td><em>selected_cds_exons_fraction</em></td>
<td>Returns the fraction of CDS segments in the selected ORF
(irrespective of the number of exons involved)</td>
<td>Float</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>selected_cds_fraction</em></td>
<td>This property calculates the fraction of the selected CDS
vs. the cDNA length.</td>
<td>Float</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>selected_cds_intron_fraction</em></td>
<td>This property returns the fraction of CDS introns of the
selected ORF of the transcript vs. the total number of
CDS introns in the Locus (considering only the selected
ORF). If the transcript is by itself, it should return 1.</td>
<td>Float</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>selected_cds_length</em></td>
<td>This property calculates the length of the CDS selected
as best inside the cDNA.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>selected_cds_num</em></td>
<td>This property calculates the number of CDS exons for the
selected ORF.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>selected_cds_number_fraction</em></td>
<td>This property returns the proportion of best possible CDS
segments vs. the number of exons. See selected_cds_number.</td>
<td>Float</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>selected_end_distance_from_junction</em></td>
<td>This metric returns the distance between the stop codon
and the last junction of the transcript. In many
eukaryotes, this distance cannot exceed 50-55 bps,
otherwise the transcript becomes a target for NMD. If the
transcript is not coding or there is no junction
downstream of the stop codon, the metric returns 0.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>selected_end_distance_from_tes</em></td>
<td>This property returns the distance of the end of the best
CDS from the transcript end site. If no CDS is defined,
it defaults to 0.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>selected_start_distance_from_tss</em></td>
<td>This property returns the distance of the start of the
best CDS from the transcript start site. If no CDS is
defined, it defaults to 0.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-even"><td><em>snowy_blast_score</em></td>
<td>Metric that indicates how good a hit is compared to the
competition, in terms of BLAST similarities. As in
SnowyOwl <a class="reference internal" href="References.html#snowyowl" id="id3">[SnowyOwl]</a>, the score for each hit is calculated
by taking the percentage of positive matches and dividing
it by (2 * len(self.blast_hits)). IMPORTANT: when
splitting transcripts by ORF, a blast hit is added to the
new transcript only if it is contained within it. This
will influnce directly this metric.</td>
<td>Float</td>
<td><strong>External</strong></td>
</tr>
<tr class="row-odd"><td><em>source_score</em></td>
<td>This metric returns a score that is assigned to the
transcript solely in virtue of its origin.</td>
<td>Float</td>
<td><strong>External</strong></td>
</tr>
<tr class="row-even"><td><em>start_distance_from_tss</em></td>
<td>This property returns the distance of the start of the
combined CDS from the transcript start site.
If no CDS is defined, it defaults to 0.</td>
<td>Int</td>
<td><strong>CDS</strong></td>
</tr>
<tr class="row-odd"><td><em>three_utr_length</em></td>
<td>Returns the length of the 5&#8217; UTR of the selected ORF.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-even"><td><em>three_utr_num</em></td>
<td>This property returns the number of 3&#8217; UTR segments
(referred to the selected ORF).</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-odd"><td><em>three_utr_num_complete</em></td>
<td>This property returns the number of 3&#8217; UTR segments for
the selected ORF, considering only those which are
complete exons.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-even"><td><em>utr_fraction</em></td>
<td>This property calculates the length of the UTR of the
selected ORF vs. the cDNA length.</td>
<td>Float</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-odd"><td><em>utr_length</em></td>
<td>Returns the sum of the 5&#8217;+3&#8217; UTR lengths.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-even"><td><em>utr_num</em></td>
<td>Returns the number of UTR segments.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-odd"><td><em>utr_num_complete</em></td>
<td>Returns the number of UTR segments which are complete
exons.</td>
<td>Int</td>
<td><strong>UTR</strong></td>
</tr>
<tr class="row-even"><td><em>verified_introns_num</em></td>
<td>This metric returns the number of introns of the
transcript which are validated by external data.</td>
<td>Int</td>
<td><strong>External</strong></td>
</tr>
</tbody>
</table>
</div>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Installation.html" class="btn btn-neutral float-right" title="2. Installation" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="Mikado: pick your transcript" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Venturini Luca, Caim Shabhonam, Mapleson Daniel, Kaithakottil Gemy George, Swarbreck David.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0b4',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>