<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>1. Introduction &mdash; Mikado 1.0.0b4 documentation</title>
    
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0b4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="Mikado 1.0.0b4 documentation" href="index.html" />
    <link rel="next" title="2. Installation" href="Installation.html" />
    <link rel="prev" title="Mikado: pick your transcript" href="index.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Installation.html" title="2. Installation"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Mikado: pick your transcript"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Mikado 1.0.0b4 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <span class="target" id="introduction"></span><div class="section" id="id1">
<h1>1. Introduction<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h1>
<p><a class="reference external" href="https://www.python.org/"><img alt="python_badge" src="https://img.shields.io/pypi/pyversions/snakemake.svg?style=flat-square" /></a> <a class="reference external" href="http://snakemake.bitbucket.org"><img alt="snake_badge" src="https://img.shields.io/badge/snakemake-≥3.5.2-brightgreen.svg?style=flat-square)]" /></a></p>
<dl class="docutils">
<dt>Numerous algorithms have been proposed to analyse RNA-Seq data, both in terms of aligning the reads to a reference genome (<a class="reference internal" href="References.html#tophat2" id="id2">[TopHat2]</a>, <a class="reference internal" href="References.html#star" id="id3">[STAR]</a>, <a class="reference internal" href="References.html#hisat" id="id4">[Hisat]</a>) or to assemble them to infer the sequence and structure of the original molecules present in the sample. The latter phase can be performed either by using alignment data against the reference genome (<a class="reference internal" href="References.html#cufflinks" id="id5">[Cufflinks]</a>, <a class="reference internal" href="References.html#stringtie" id="id6">[StringTie]</a>, <a class="reference internal" href="References.html#class2" id="id7">[Class2]</a>, <a class="reference internal" href="References.html#trinity" id="id8">[Trinity]</a>) or in the absence of such information (<a class="reference internal" href="References.html#trinity" id="id9">[Trinity]</a>, <a class="reference internal" href="References.html#oases" id="id10">[Oases]</a>, <a class="reference internal" href="References.html#bridger" id="id11">[Bridger]</a>, ). Each of these methods has to contend with numerous sources of variability in the RNA-Seq data:</dt>
<dd><ul class="first last simple">
<li>alternative splicing events at the same locus</li>
<li>extremely variable expression and therefore sequencing depth at different loci</li>
<li>presence of orthologous genes with very similar cDNAs.</li>
<li>tandem duplicated genes which can be artefactually reconstructed as a single, fused entity</li>
</ul>
</dd>
</dl>
<p>Multiple assessments of these methods on real and simulated data. In particular, the RGASP consortium promoted a competition among some of the most popular methods in 2012 <a class="reference internal" href="References.html#rgasp" id="id12">[RGASP]</a>; the results showed that the accuracy of the methods was dependent on the input data, the expression levels of the genes, and on the species under analysis. A different concern regards the completeness of each of the assembly methods, ie whether methods that are not necessarily the best across all gene loci might nonetheless be capable of outperforming other competitors for specific contigs. Recently, approaches have been proposed to integrate multiple RNA-Seq assemblies by inferring their protein-coding content and collapsing them accordingly <a class="reference internal" href="References.html#evigenetobacco" id="id13">[EviGeneTobacco]</a> or to determine the best proposed assembly using multiple measures related to how the transcript compares to the input RNA-Seq reads (<a class="reference internal" href="References.html#transrate" id="id14">[Transrate]</a>, <a class="reference internal" href="References.html#rsemeval" id="id15">[RSEMeval]</a>).</p>
<p>A different but related line of research has focused on how to integrate data coming from multiple samples. While some researchers advocate for merging together the input data and assembling it <a class="reference internal" href="References.html#trinity" id="id16">[Trinity]</a>, others have developed methods to integrate multiple assemblies into a single coherent annotation (<a class="reference internal" href="References.html#cuffmerge" id="id17">[CuffMerge]</a>, Stringtie-merge). Another common approach relies on a third party tool to integrate multiple assemblies together with other data - typically protein alignments and <em>ab initio</em> predictions - to select for the best model. The latter approach has been chosen by some of the most popular pipelines for genome annotation in the last years (<a class="reference internal" href="References.html#evm" id="id18">[EVM]</a>, <a class="reference internal" href="References.html#maker" id="id19">[Maker]</a>).</p>
<p>Our tool, Mikado, contributes to this area of research by proposing a novel algorithm to integrate multiple transcript assemblies, leveraging additional data regarding the position of ORFs (generally derived with Transdecoder <a class="reference internal" href="References.html#trinity" id="id20">[Trinity]</a>), sequence similarity (derived with BLAST <a class="reference internal" href="References.html#blastplus" id="id21">[Blastplus]</a>) and reliable splicing junctions (generally derived using Portcullis <a class="reference internal" href="References.html#portcullis" id="id22">[Portcullis]</a>). Through the combination of these input data sources, we are capable of identifying artefacts - especially gene fusions and small fragments - and retrieve the original transcripts in each locus with high accuracy and sensitivity.</p>
<p>An important caveat is that our approach does not look for the <em>best</em> candidate in terms of the input data, as Transrate <a class="reference internal" href="References.html#transrate" id="id23">[Transrate]</a> or RSEM-Eval <a class="reference internal" href="References.html#rsemeval" id="id24">[RSEMeval]</a> do. Rather, our approach is more similar to EvidentialGene <a class="reference internal" href="References.html#evigenetobacco" id="id25">[EviGeneTobacco]</a> as Mikado will try to find the candidates most likely to represent the isoforms which would be annotated as canonical by a human annotator. Biological events such as intron retentions or trans-splicing that might be present in the sample are explicitly selected against, in order to provide a clean annotation of the genome.</p>
<div class="section" id="overview">
<h2>1.1. Overview<a class="headerlink" href="#overview" title="Permalink to this headline">¶</a></h2>
<p>Mikado analyses an ensemble of RNA-Seq assemblies by looking for overlapping transcripts based on their genomic position. Such clusters of genes, or <em>superloci</em>, are then analysed to locate the gene loci that will form the final annotation. In order, the steps of the pipeline are as follows:</p>
<ol class="arabic simple">
<li><a class="reference internal" href="Usage/Prepare.html#prepare"><span class="std std-ref">In the first step</span></a>, <code class="docutils literal"><span class="pre">mikado</span> <span class="pre">prepare</span></code> will combine assemblies from multiple sources into a single coherent, filtered dataset.</li>
<li><a class="reference internal" href="Usage/Serialise.html#serialise"><span class="std std-ref">In the second step</span></a>, <code class="docutils literal"><span class="pre">mikado</span> <span class="pre">serialise</span></code> will collect data from multiple sources (<a class="reference external" href="https://github.com/maplesond/portcullis">Portcullis</a>, <a class="reference external" href="http://transdecoder.github.io/">Transdecoder</a>, BLASTX) and load it into a database for fast integration with the input data.</li>
<li><a class="reference internal" href="Usage/Pick.html#pick"><span class="std std-ref">In the final step</span></a>, <code class="docutils literal"><span class="pre">mikado</span> <span class="pre">pick</span></code> will integrate the prepared transcripts with the serialised additional data to choose the best transcripts.</li>
</ol>
<p>By far, the most important source of data for Mikado is <a class="reference external" href="http://transdecoder.github.io/">Transdecoder</a>, as the algorithms for finding and splitting chimeras depend on the detection of the ORFs on the transcript sequences. Please refer to the <a class="reference internal" href="Algorithms.html#algorithms"><span class="std std-ref">Algorithms</span></a> section for more details on how Mikado operates.</p>
</div>
<div class="section" id="the-command-line-interface">
<h2>1.2. The command-line interface<a class="headerlink" href="#the-command-line-interface" title="Permalink to this headline">¶</a></h2>
<p>The Mikado suite provides two commands: <code class="docutils literal"><span class="pre">mikado</span></code> and <code class="docutils literal"><span class="pre">daijin</span></code>. The former provides access to the main functionalities of the suite:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ mikado --help
usage: Mikado [-h] {configure,prepare,serialise,pick,compare,util} ...

Mikado is a program to analyse RNA-Seq data and determine the best transcript
for each locus in accordance to user-specified criteria.

optional arguments:
  -h, --help            show this help message and exit

Components:
  {configure,prepare,serialise,pick,compare,util}
                        These are the various components of Mikado:
    configure           This utility guides the user through the process of
                        creating a configuration file for Mikado.
    prepare             Mikado prepare analyses an input GTF file and prepares
                        it for the picking analysis by sorting its transcripts
                        and performing some simple consistency checks.
    serialise           Mikado serialise creates the database used by the pick
                        program. It handles Junction and ORF BED12 files as
                        well as BLAST XML results.
    pick                Mikado pick analyses a sorted GTF/GFF files in order
                        to identify its loci and choose the best transcripts
                        according to user-specified criteria. It is dependent
                        on files produced by the &quot;prepare&quot; and &quot;serialise&quot;
                        components.
    compare             Mikado compare produces a detailed comparison of
                        reference and prediction files. It has been directly
                        inspired by Cufflinks&#39;s cuffcompare and ParsEval.
    util                Miscellaneous utilities
</pre></div>
</div>
<p>Each of these subcommands is explained in detail in the <span class="xref std std-ref">Usage</span> section.</p>
<p><code class="docutils literal"><span class="pre">daijin</span></code> instead provides the interface to the <a href="#id26"><span class="problematic" id="id27">:ref:`Daijin pipeline manager &lt;Daijin&gt;`_</span></a>, which manages the task of going from a dataset of multiple reads to the Mikado final picking. This is its interface:</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>$ daijin --help

usage: A Directed Acyclic pipeline <span class="k">for</span> gene model reconstruction from RNA seq data.
        Basically, a pipeline <span class="k">for</span> driving Mikado. It will first align RNAseq reads against
        a genome using multiple tools, <span class="k">then</span> creates transcript assemblies using multiple tools,
        and find junctions in the alignments using Portcullis.
        This input is <span class="k">then</span> passed into Mikado.
       <span class="o">[</span>-h<span class="o">]</span> <span class="o">{</span>configure,assemble,mikado<span class="o">}</span> ...

optional arguments:
  -h, --help            show this <span class="nb">help</span> message and <span class="nb">exit</span>

Pipelines:
  <span class="o">{</span>configure,assemble,mikado<span class="o">}</span>
                        These are the pipelines that can be executed via
                        daijin.
    configure           Creates the configuration files <span class="k">for</span> Daijin execution.
    assemble            A pipeline that generates a variety of transcript
                        assemblies using various aligners and assemblers, as
                        well a producing a configuration file suitable <span class="k">for</span>
                        driving Mikado.
    mikado              Using a supplied configuration file that describes all
                        input assemblies to use, it runs the Mikado pipeline,
                        including prepare, BLAST, transdecoder, serialise and
                        pick.
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/mikado-logo.jpg" alt="Logo"/>
            </a></p>
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">1. Introduction</a><ul>
<li><a class="reference internal" href="#overview">1.1. Overview</a></li>
<li><a class="reference internal" href="#the-command-line-interface">1.2. The command-line interface</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Mikado: pick your transcript</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="Installation.html"
                        title="next chapter">2. Installation</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/Introduction.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="Installation.html" title="2. Installation"
             >next</a> |</li>
        <li class="right" >
          <a href="index.html" title="Mikado: pick your transcript"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">Mikado 1.0.0b4 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Venturini Luca, Caim Shabhonam, Mapleson Daniel, Kaithakottil Gemy George, Swarbreck David.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.5.
    </div>
  </body>
</html>