

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>4. References &mdash; Mikado 1.0.0b4 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  
    <link rel="stylesheet" href="_static/theme_overrides.css" type="text/css" />
  

  
    <link rel="top" title="Mikado 1.0.0b4 documentation" href="index.html"/>
        <link rel="next" title="Library documentation" href="Library/index.html"/>
        <link rel="prev" title="3. Mikado miscellaneous scripts" href="Usage_dir/Utilities.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Mikado
          

          
            
            <img src="_static/mikado-logo.jpg" class="logo" />
          
          </a>

          
            
            
              <div class="version">
                1.0
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
                <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="Introduction.html">1. Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="Installation.html">2. Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="Tutorial.html">3. Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="Usage.html">4. Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="Usage.html#mikado-pipeline-stages">5. Mikado pipeline stages</a></li>
<li class="toctree-l1"><a class="reference internal" href="Usage.html#mikado-utilities">6. Mikado utilities</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">4. References</a></li>
<li class="toctree-l1"><a class="reference internal" href="Library/index.html">Library documentation</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="index.html">Mikado</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          

 



<div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="index.html">Docs</a> &raquo;</li>
      
    <li>4. References</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/References.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="references">
<h1>4. References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h1>
<table class="docutils citation" frame="void" id="cufflinks" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Cufflinks]</td><td><strong>Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation</strong> Cole Trapnell, Brian Williams, Geo Pertea, Ali Mortazavi, Gordon Kwan, Jeltje van Baren, Steven Salzberg, Barbara Wold, Lior Pachter. <em>Nature Biotechnology</em>, 2010, doi:10.1038/nbt.1621</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="parseval" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[ParsEval]</td><td><strong>ParsEval: parallel comparison and analysis of gene structure annotations</strong> Daniel S Standage and Volker P Brendel. <em>BMC Bioinformatics</em>, 2012, doi:10.1186/1471-2105-13-187</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="rgasp" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[RGASP]</td><td><strong>Assessment of transcript reconstruction methods for RNA-seq</strong>  Tamara Steijger, Josep F Abril, Pär G Engström, Felix Kokocinski, The RGASP Consortium, Tim J Hubbard, Roderic Guigó, Jennifer Harrow &amp; Paul Bertone. <em>Nature Methods</em>, 2013, doi:10.1038/nmeth.2714</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="snowyowl" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[SnowyOwl]</td><td><strong>SnowyOwl: accurate prediction of fungal genes by using RNA-Seq and homology information to select among ab initio models</strong> Ian Reid, Nicholas O’Toole, Omar Zabaneh, Reza Nourzadeh, Mahmoud Dahdouli, Mostafa Abdellateef, Paul MK Gordon, Jung Soh, Gregory Butler, Christoph W Sensen and Adrian Tsang. <em>BMC Bioinformatics</em>, 2014, doi:10.1186/1471-2105-15-229</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="cuffmerge" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[CuffMerge]</td><td><strong>Identification of novel transcripts in annotated genomes using RNA-Seq</strong> Adam Roberts, Harold Pimentel, Cole Trapnell and Lior Pachter. <em>Bioinformatics</em>, 2011, doi:10.1093/bioinformatics/btr355</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="class2" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Class2]</td><td><strong>CLASS2: accurate and efficient splice variant annotation from RNA-seq reads</strong> Li Song, Sarven Sabunciyan and Liliana Florea. <em>Bioinformatics</em>, 2016, doi:10.1093/nar/gkw158</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="pyfaidx" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[PyFaidx]</td><td><strong>Efficient &#8220;pythonic&#8221; access to FASTA files using pyfaidx</strong> Matthew D Shirley​, Zhaorong Ma, Brent S Pedersen and Sarah J Wheelan. <em>PeerJ PrePrints</em> 3:e1196, 2015. doi:10.7287/peerj.preprints.970v1</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="snake" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Snake]</td><td><strong>Snakemake—a scalable bioinformatics workflow engine</strong> Johannes Köster and Sven Rahmann1. <em>Bioinformatics</em>, 2012, doi:10.1093/bioinformatics/bts480</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="trinity" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Trinity]</td><td><strong>De novo transcript sequence reconstruction from RNA-Seq: reference generation and analysis with Trinity</strong> Brian J Haas, et al. <em>Nature Protocols</em>, 2013. doi:10.1038/nprot.2013.084</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="blastplus" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Blastplus]</td><td><strong>BLAST+: architecture and applications.</strong> Camacho C, Coulouris G, Avagyan V, Ma N, Papadopoulos J, Bealer K, Madden TL. <em>BMC Bioinformatics</em>, 2009. doi:10.1186/1471-2105-10-421</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="star" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[STAR]</td><td><strong>STAR: ultrafast universal RNA-seq aligner</strong> Alexander Dobin, Carrie A. Davis1, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha1, Philippe Batut1, Mark Chaisson and Thomas R. Gingeras. <em>Bioinformatics</em>, 2012. doi:10.1093/bioinformatics/bts635</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="hisat" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Hisat]</td><td><strong>HISAT: a fast spliced aligner with low memory requirements</strong> Daehwan Kim, Ben Langmead and Stevan L Salzberg. <em>Nature Methods</em>, 2015. doi:10.1038/nmeth.3317</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="tophat2" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[TopHat2]</td><td><strong>TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions</strong> Daehwan Kim, Geo Pertea, Cole Trapnell, Harold Pimentel, Ryan Kelley and Steven L Salzberg. <em>Genome Biology</em>, 2013. doi:10.1186/gb-2013-14-4-r36</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="stringtie" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[StringTie]</td><td><strong>StringTie enables improved reconstruction of a transcriptome from RNA-seq reads</strong>  Mihaela Pertea, Geo M Pertea, Corina M Antonescu, Tsung-Cheng Chang, Joshua T Mendell       and Steven L Salzberg. <em>Nature Biotechnology</em>, 2015. doi:10.1038/nbt.3122</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="gmap" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[GMAP]</td><td><strong>GMAP: a genomic mapping and alignment program for mRNA and EST sequences</strong> Thomas D. Wu and Colin K. Watanabe. <em>Bioinformatics</em> 2005. doi:10.1093/bioinformatics/bti310</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="uorfs" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[uORFs]</td><td><strong>uAUG and uORFs in human and rodent 5′untranslated mRNAs.</strong> Michele Iacono, Flavio Mignone and Graziano Pesole. <em>Gene</em>, 2005. doi:10.1016/j.gene.2004.11.041</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="pyyaml" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[PyYaml]</td><td><strong>Pyyaml library</strong> K Simonov. <em>http://pyyaml.org/</em>, 2006.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="cython" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Cython]</td><td><strong>Cython: the best of both worlds.</strong> Stefan Behnel, RObert Bradshaw, Craig Citro, Lisandro Dalcin, Dag Sverre Seljebotn and Kurt Smith. <em>AIP - Computing in science &amp; engineering</em>, 2011. doi:10.1109/MCSE.2010.118</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="numpy" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Numpy]</td><td><strong>The NumPy array: A structure for efficient numerical computation.</strong> Stefan van der Walt, S. Chris Colbert and Gael Varoquaux. <em>Computing in Science &amp; Engineering</em>, 2011. doi:10.1109/MCSE.2011.37</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="scipy" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[Scipy]</td><td><strong>SciPy: Open Source Scientific Tools for Python.</strong> Eric Jones and Travis Oliphant and Pearu Peterson et al. <em>http://www.scipy.org/*</em>, 2001.</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="networkx" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[NetworkX]</td><td><strong>Exploring network structure, dynamics, and function using NetworkX</strong> Aric A. Hagberg, Daniel A. Schult and Pieter J. Swart. <em>Proceedings of the 7th Python in Science Conference (SciPy2008)</em>, 2008. doi:</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="biopython" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[BioPython]</td><td><strong>Biopython: freely available Python tools for computational molecular biology and bioinformatics.</strong> PA Cock, T Antao, JT Chang, BA Bradman, CJ Cox, A Dalke, I Friedberg, T Hamelryck, F Kauff, B Wilczynski and MJL de Hoon. <em>Bioinformatics</em>, 2009. doi:10.1093/bioinformatics/btp163</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="scikit" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[SciKit]</td><td><strong>Scikit-learn: Machine Learning in Python.</strong> Fabian Pedregosa, Gael Varoquaux, Alexandre Gramfort, Vincent Michel, Bertrand Thirion, Olivier Grisel, Mathieu Blondel, Peter Prettenhofer, Ron Weiss, Vincent Dubourg, Jake Vanderplas, Alexandre Passos, David Cournapeau, Matthieu Brucher, Matthieu Perrot and Édouard Duchesnay. <em>The Journal of Machine Learning Research</em>, 2011. doi:</td></tr>
</tbody>
</table>
<table class="docutils citation" frame="void" id="drmaa" rules="none">
<colgroup><col class="label" /><col /></colgroup>
<tbody valign="top">
<tr><td class="label">[DRMAA]</td><td><strong>Distributed resource management application API Version 2 (DRMAA).</strong> P Tröger, R Brobst, D Gruber, M Mamonski and D Templeton. <em>Open Grid Forum</em>, 2012. doi:</td></tr>
</tbody>
</table>
</div>


           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="Library/index.html" class="btn btn-neutral float-right" title="Library documentation" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="Usage_dir/Utilities.html" class="btn btn-neutral" title="3. Mikado miscellaneous scripts" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2016, Venturini Luca, Caim Shabhonam, Mapleson Daniel, Kaithakottil Gemy George, Swarbreck David.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.0.0b4',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>