References
==========

.. [Cufflinks] **Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation** Cole Trapnell, Brian Williams, Geo Pertea, Ali Mortazavi, Gordon Kwan, Jeltje van Baren, Steven Salzberg, Barbara Wold, Lior Pachter. *Nature Biotechnology*, 2010, doi:10.1038/nbt.1621
.. [ParsEval] **ParsEval: parallel comparison and analysis of gene structure annotations** Daniel S Standage and Volker P Brendel. *BMC Bioinformatics*, 2012, doi:10.1186/1471-2105-13-187
.. [RGASP] **Assessment of transcript reconstruction methods for RNA-seq**  Tamara Steijger, Josep F Abril, Pär G Engström, Felix Kokocinski, The RGASP Consortium, Tim J Hubbard, Roderic Guigó, Jennifer Harrow & Paul Bertone. *Nature Methods*, 2013, doi:10.1038/nmeth.2714
.. [SnowyOwl] **SnowyOwl: accurate prediction of fungal genes by using RNA-Seq and homology information to select among ab initio models** Ian Reid, Nicholas O’Toole, Omar Zabaneh, Reza Nourzadeh, Mahmoud Dahdouli, Mostafa Abdellateef, Paul MK Gordon, Jung Soh, Gregory Butler, Christoph W Sensen and Adrian Tsang. *BMC Bioinformatics*, 2014, doi:10.1186/1471-2105-15-229
.. [CuffMerge] **Identification of novel transcripts in annotated genomes using RNA-Seq** Adam Roberts, Harold Pimentel, Cole Trapnell and Lior Pachter. *Bioinformatics*, 2011, doi:10.1093/bioinformatics/btr355
.. [Class2] **CLASS2: accurate and efficient splice variant annotation from RNA-seq reads** Li Song, Sarven Sabunciyan and Liliana Florea. *Bioinformatics*, 2016, doi:10.1093/nar/gkw158
.. [PyFaidx] **Efficient "pythonic" access to FASTA files using pyfaidx** Matthew D Shirley​, Zhaorong Ma, Brent S Pedersen and Sarah J Wheelan. *PeerJ PrePrints* 3:e1196, 2015. doi:10.7287/peerj.preprints.970v1
.. [Snake] **Snakemake—a scalable bioinformatics workflow engine** Johannes Köster and Sven Rahmann1. *Bioinformatics*, 2012, doi:10.1093/bioinformatics/bts480
.. [Trinity] **De novo transcript sequence reconstruction from RNA-Seq: reference generation and analysis with Trinity** Brian J Haas, et al. *Nature Protocols*, 2013. doi:10.1038/nprot.2013.084
.. [Blastplus] **BLAST+: architecture and applications.** Camacho C, Coulouris G, Avagyan V, Ma N, Papadopoulos J, Bealer K, Madden TL. *BMC Bioinformatics*, 2009. doi:10.1186/1471-2105-10-421
.. [STAR] **STAR: ultrafast universal RNA-seq aligner** Alexander Dobin, Carrie A. Davis1, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha1, Philippe Batut1, Mark Chaisson and Thomas R. Gingeras. *Bioinformatics*, 2012. doi:10.1093/bioinformatics/bts635
.. [Hisat] **HISAT: a fast spliced aligner with low memory requirements** Daehwan Kim, Ben Langmead and Stevan L Salzberg. *Nature Methods*, 2015. doi:10.1038/nmeth.3317
.. [TopHat2] **TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions** Daehwan Kim, Geo Pertea, Cole Trapnell, Harold Pimentel, Ryan Kelley and Steven L Salzberg. *Genome Biology*, 2013. doi:10.1186/gb-2013-14-4-r36
.. [StringTie] **StringTie enables improved reconstruction of a transcriptome from RNA-seq reads**  Mihaela Pertea, Geo M Pertea, Corina M Antonescu, Tsung-Cheng Chang, Joshua T Mendell	and Steven L Salzberg. *Nature Biotechnology*, 2015. doi:10.1038/nbt.3122
.. [GMAP] **GMAP: a genomic mapping and alignment program for mRNA and EST sequences** Thomas D. Wu and Colin K. Watanabe. *Bioinformatics* 2005. doi:10.1093/bioinformatics/bti310