References
==========

.. [Mikado] **Leveraging multiple transcriptome assembly methods for improved gene structure annotation** Luca Venturini, Shabhonam Caim, Gemy George Kaithakottil, Daniel Lee Mapleson, David Swabreck. *GigaScience*, 2018. doi: 10.1093/gigascience/giy093
.. [Cufflinks]  **Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation** Cole Trapnell, Brian Williams, Geo Pertea, Ali Mortazavi, Gordon Kwan, Jeltje van Baren, Steven Salzberg, Barbara Wold, Lior Pachter. *Nature Biotechnology*, 2010, doi:10.1038/nbt.1621
.. [ParsEval]  **ParsEval: parallel comparison and analysis of gene structure annotations** Daniel S Standage and Volker P Brendel. *BMC Bioinformatics*, 2012, doi:10.1186/1471-2105-13-187
.. [RGASP]  **Assessment of transcript reconstruction methods for RNA-seq**  Tamara Steijger, Josep F Abril, Pär G Engström, Felix Kokocinski, The RGASP Consortium, Tim J Hubbard, Roderic Guigó, Jennifer Harrow & Paul Bertone. *Nature Methods*, 2013, doi:10.1038/nmeth.2714
.. [SnowyOwl]  **SnowyOwl: accurate prediction of fungal genes by using RNA-Seq and homology information to select among ab initio models** Ian Reid, Nicholas O’Toole, Omar Zabaneh, Reza Nourzadeh, Mahmoud Dahdouli, Mostafa Abdellateef, Paul MK Gordon, Jung Soh, Gregory Butler, Christoph W Sensen and Adrian Tsang. *BMC Bioinformatics*, 2014, doi:10.1186/1471-2105-15-229
.. [CuffMerge]  **Identification of novel transcripts in annotated genomes using RNA-Seq** Adam Roberts, Harold Pimentel, Cole Trapnell and Lior Pachter. *Bioinformatics*, 2011, doi:10.1093/bioinformatics/btr355
.. [Class2]  **CLASS2: accurate and efficient splice variant annotation from RNA-seq reads** Li Song, Sarven Sabunciyan and Liliana Florea. *Bioinformatics*, 2016, doi:10.1093/nar/gkw158
.. [PyFaidx]  **Efficient "pythonic" access to FASTA files using pyfaidx** Matthew D Shirley​, Zhaorong Ma, Brent S Pedersen and Sarah J Wheelan. *PeerJ PrePrints* 3:e1196, 2015. doi:10.7287/peerj.preprints.970v1
.. [Snake]  **Snakemake—a scalable bioinformatics workflow engine** Johannes Köster and Sven Rahmann1. *Bioinformatics*, 2012, doi:10.1093/bioinformatics/bts480
.. [Trinity]  **De novo transcript sequence reconstruction from RNA-Seq: reference generation and analysis with Trinity** Brian J Haas, et al. *Nature Protocols*, 2013. doi:10.1038/nprot.2013.084
.. [Blastplus]  **BLAST+: architecture and applications.** Camacho C, Coulouris G, Avagyan V, Ma N, Papadopoulos J, Bealer K, Madden TL. *BMC Bioinformatics*, 2009. doi:10.1186/1471-2105-10-421
.. [Diamond] **Fast and sensitive protein alignment using DIAMOND** B Buchfink, C Xie, D H Huson. *Nature Methods* 12, 59-60 (2015). doi:10.1038/nmeth.3176
.. [STAR]  **STAR: ultrafast universal RNA-seq aligner** Alexander Dobin, Carrie A. Davis1, Felix Schlesinger, Jorg Drenkow, Chris Zaleski, Sonali Jha1, Philippe Batut1, Mark Chaisson and Thomas R. Gingeras. *Bioinformatics*, 2012. doi:10.1093/bioinformatics/bts635
.. [Hisat]  **HISAT: a fast spliced aligner with low memory requirements** Daehwan Kim, Ben Langmead and Stevan L Salzberg. *Nature Methods*, 2015. doi:10.1038/nmeth.3317
.. [TopHat2]  **TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions** Daehwan Kim, Geo Pertea, Cole Trapnell, Harold Pimentel, Ryan Kelley and Steven L Salzberg. *Genome Biology*, 2013. doi:10.1186/gb-2013-14-4-r36
.. [StringTie]  **StringTie enables improved reconstruction of a transcriptome from RNA-seq reads**  Mihaela Pertea, Geo M Pertea, Corina M Antonescu, Tsung-Cheng Chang, Joshua T Mendell	and Steven L Salzberg. *Nature Biotechnology*, 2015. doi:10.1038/nbt.3122
.. [GMAP]  **GMAP: a genomic mapping and alignment program for mRNA and EST sequences** Thomas D. Wu and Colin K. Watanabe. *Bioinformatics* 2005. doi:10.1093/bioinformatics/bti310
.. [uORFs]  **uAUG and uORFs in human and rodent 5′untranslated mRNAs.** Michele Iacono, Flavio Mignone and Graziano Pesole. *Gene*, 2005. doi:10.1016/j.gene.2004.11.041
.. [PyYaml]  **Pyyaml library** K Simonov. *http://pyyaml.org/*, 2006.
.. [Cython]  **Cython: the best of both worlds.** Stefan Behnel, RObert Bradshaw, Craig Citro, Lisandro Dalcin, Dag Sverre Seljebotn and Kurt Smith. *AIP - Computing in science & engineering*, 2011. doi:10.1109/MCSE.2010.118
.. [Numpy]  **The NumPy array: A structure for efficient numerical computation.** Stefan van der Walt, S. Chris Colbert and Gael Varoquaux. *Computing in Science & Engineering*, 2011. doi:10.1109/MCSE.2011.37
.. [Scipy]  **SciPy: Open Source Scientific Tools for Python.** Eric Jones and Travis Oliphant and Pearu Peterson et al. *http://www.scipy.org/**, 2001.
.. [NetworkX]  **Exploring network structure, dynamics, and function using NetworkX** Aric A. Hagberg, Daniel A. Schult and Pieter J. Swart. *Proceedings of the 7th Python in Science Conference (SciPy2008)*, 2008. doi:
.. [BioPython]  **Biopython: freely available Python tools for computational molecular biology and bioinformatics.** PA Cock, T Antao, JT Chang, BA Bradman, CJ Cox, A Dalke, I Friedberg, T Hamelryck, F Kauff, B Wilczynski and MJL de Hoon. *Bioinformatics*, 2009. doi:10.1093/bioinformatics/btp163
.. [DRMAA]  **Distributed resource management application API Version 2 (DRMAA).** P Tröger, R Brobst, D Gruber, M Mamonski and D Templeton. *Open Grid Forum*, 2012. doi:
.. [Apollo]  **Web Apollo: a web-based genomic annotation editing platform** Eduardo Lee, Gregg A Helt, Justin T Reese, Monica C Munoz-Torres, Chris P Childers, Robert M Buels, Lincoln Stein, Ian H Holmes, Christine G Elsik and Suzanna E Lewis. *Genome Biology*, 2013, doi:10.1186/gb-2013-14-8-r93
.. [Rampart]  **RAMPART: a workflow management system for de novo genome assembly**, Daniel Mapleson, Nizar Drou and David Swarbreck. *Bioinformatics*, 2015. doi:10.1093/bioinformatics/btv056
.. [Uniprot]  **UniProt: a hub for protein information** The UniProt Consortium. *Nucleic Acid Research*, 2014. doi:10.1093/nar/gku989
.. [Portcullis]  **Efficient and accurate detection of splice junctions from RNA-seq with Portcullis** Daniel Lee Mapleson, Luca Venturini, Gemy George Kaithakottil, David Swarbreck. *GigaScience*, 2018. doi: 10.1093/gigascience/giy131
.. [Oases]  **Oases: robust de novo RNA-seq assembly across the dynamic range of expression levels** Marcel H Schulz, Daniel R Zerbino, Martin Vingron and Ewan Birney. *Bioinformatics*, 2012. doi: 10.1093/bioinformatics/bts094
.. [Bridger]  **Bridger: a new framework for de novo transcriptome assembly using RNA-seq data** Zheng Chang, Guojun Li, Juntao Liu, Cody Ashby, Deli Liu, Carole L Cramer and Xiuzhen Huang. *Genome Biology*, 2015. doi:10.1186/s13059-015-0596-2
.. [EviGeneTobacco]  **Combining Transcriptome Assemblies from Multiple De Novo Assemblers in the Allo-Tetraploid Plant Nicotiana benthamiana** Kenlee Nakasugi, Ross Crowhurst, Julia Bally, Peter Waterhouse. *PLoS ONE*, 2014. doi:10.1371/journal.pone.0091776
.. [TransAbyss]  **De novo assembly and analysis of RNA-seq data** Gordon Robertson et al., *Nature Methods*, 2010. doi:10.1038/nmeth.1517
.. [Transrate]  **TransRate: reference-free quality assessment of de novo transcriptome assemblies** Richard Smith-Unna, Chris Boursnell, Rob Patro, Julian M. Hibberd and Steven Kelly. *Genome Research*, 2016. doi:10.1101/gr.196469.115
.. [RSEMeval]  **Evaluation of de novo transcriptome assemblies from RNA-Seq data** Bo Li, Nathanael Fillmore, Yongsheng Bai, Mike Collins, James A Thomson, Ron Stewart and Colin N Dewey. *Genome Biology*, 2014. doi:10.1186/s13059-014-0553-5
.. [Maker]  **MAKER2: an annotation pipeline and genome-database management tool for second-generation genome projects** Carson Holt and Mark Yandell. *BMC Bioinformatics*, 2011. doi:10.1186/1471-2105-12-491
.. [EVM]  **Automated eukaryotic gene structure annotation using EVidenceModeler and the Program to Assemble Spliced Alignments** Brian J Haas, Steven L Salzberg, Wei Zhu, Mihaela Pertea, Jonathan E Allen, Joshua Orvis, Owen White, C Robin Buell and Jennifer R Wortman. *Genome Biology*, 2008. doi:10.1186/gb-2008-9-1-r7
.. [Augustus]  **WebAUGUSTUS—a web service for training AUGUSTUS and predicting genes in eukaryotes** Katharina J. Hoff and Mario Stanke. *Nucleic Acid Research*, 2013. doi:10.1093/nar/gkt418
.. [Maker2]  **MAKER2: an annotation pipeline and genome-database management tool for second-generation genome projects** Carson Holt and Mark Yandell. *BMC Bioinformatics*, 2011. doi:10.1186/1471-2105-12-491
.. [AraPort]  **Araport11: a complete reannotation of the Arabidopsis thaliana reference genome** Chia‐Yi Cheng, Vivek Krishnakumar, Agnes P. Chan, Françoise Thibaud‐Nissen, Seth Schobel Christopher and D. Town. *The Plant Journal*, 2017, Volume 89, Issue 4, 789-804. doi: 10.1111/tpj.13415
.. [PYinterval]  https://github.com/chaimleib/intervaltree
.. [BXPython]  https://bitbucket.org/james_taylor/bx-python/overview
.. [Snakeviz]  https://jiffyclub.github.io/snakeviz/
.. [PASA]  **Improving the Arabidopsis genome annotation using maximal transcript alignment assemblies** Haas, B.J., Delcher, A.L., Mount, S.M., Wortman, J.R., Smith Jr, R.K., Jr., Hannick, L.I., Maiti, R., Ronning, C.M., Rusch, D.B., Town, C.D. et al. *Nucleic Acids Res*, 2003, 31, 5654-5666. doi:10.1093/nar/gkg770